{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Goal\" data-toc-modified-id=\"Goal-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Goal</a></span></li><li><span><a href=\"#Var\" data-toc-modified-id=\"Var-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Var</a></span></li><li><span><a href=\"#Init\" data-toc-modified-id=\"Init-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Init</a></span></li><li><span><a href=\"#Selecting-genome-reps\" data-toc-modified-id=\"Selecting-genome-reps-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Selecting genome reps</a></span><ul class=\"toc-item\"><li><span><a href=\"#Writing-tables\" data-toc-modified-id=\"Writing-tables-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Writing tables</a></span><ul class=\"toc-item\"><li><span><a href=\"#N100\" data-toc-modified-id=\"N100-4.1.1\"><span class=\"toc-item-num\">4.1.1&nbsp;&nbsp;</span>N100</a></span></li><li><span><a href=\"#N500\" data-toc-modified-id=\"N500-4.1.2\"><span class=\"toc-item-num\">4.1.2&nbsp;&nbsp;</span>N500</a></span></li><li><span><a href=\"#N1000\" data-toc-modified-id=\"N1000-4.1.3\"><span class=\"toc-item-num\">4.1.3&nbsp;&nbsp;</span>N1000</a></span></li><li><span><a href=\"#N10\" data-toc-modified-id=\"N10-4.1.4\"><span class=\"toc-item-num\">4.1.4&nbsp;&nbsp;</span>N10</a></span></li><li><span><a href=\"#N250\" data-toc-modified-id=\"N250-4.1.5\"><span class=\"toc-item-num\">4.1.5&nbsp;&nbsp;</span>N250</a></span></li><li><span><a href=\"#N350\" data-toc-modified-id=\"N350-4.1.6\"><span class=\"toc-item-num\">4.1.6&nbsp;&nbsp;</span>N350</a></span></li><li><span><a href=\"#N175\" data-toc-modified-id=\"N175-4.1.7\"><span class=\"toc-item-num\">4.1.7&nbsp;&nbsp;</span>N175</a></span></li></ul></li></ul></li><li><span><a href=\"#DB-create\" data-toc-modified-id=\"DB-create-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>DB-create</a></span><ul class=\"toc-item\"><li><span><a href=\"#N1000r0\" data-toc-modified-id=\"N1000r0-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>N1000r0</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-5.1.1\"><span class=\"toc-item-num\">5.1.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-5.1.2\"><span class=\"toc-item-num\">5.1.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li></ul></li><li><span><a href=\"#DB-update\" data-toc-modified-id=\"DB-update-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>DB-update</a></span><ul class=\"toc-item\"><li><span><a href=\"#n10r1\" data-toc-modified-id=\"n10r1-6.1\"><span class=\"toc-item-num\">6.1&nbsp;&nbsp;</span>n10r1</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-6.1.1\"><span class=\"toc-item-num\">6.1.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-6.1.2\"><span class=\"toc-item-num\">6.1.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li><li><span><a href=\"#n100r1\" data-toc-modified-id=\"n100r1-6.2\"><span class=\"toc-item-num\">6.2&nbsp;&nbsp;</span>n100r1</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-6.2.1\"><span class=\"toc-item-num\">6.2.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-6.2.2\"><span class=\"toc-item-num\">6.2.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li><li><span><a href=\"#n100r2\" data-toc-modified-id=\"n100r2-6.3\"><span class=\"toc-item-num\">6.3&nbsp;&nbsp;</span>n100r2</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-6.3.1\"><span class=\"toc-item-num\">6.3.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-6.3.2\"><span class=\"toc-item-num\">6.3.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li><li><span><a href=\"#n175r1\" data-toc-modified-id=\"n175r1-6.4\"><span class=\"toc-item-num\">6.4&nbsp;&nbsp;</span>n175r1</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-6.4.1\"><span class=\"toc-item-num\">6.4.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-6.4.2\"><span class=\"toc-item-num\">6.4.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li><li><span><a href=\"#n250r1\" data-toc-modified-id=\"n250r1-6.5\"><span class=\"toc-item-num\">6.5&nbsp;&nbsp;</span>n250r1</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-6.5.1\"><span class=\"toc-item-num\">6.5.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-6.5.2\"><span class=\"toc-item-num\">6.5.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li><li><span><a href=\"#n350r1\" data-toc-modified-id=\"n350r1-6.6\"><span class=\"toc-item-num\">6.6&nbsp;&nbsp;</span>n350r1</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-6.6.1\"><span class=\"toc-item-num\">6.6.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-6.6.2\"><span class=\"toc-item-num\">6.6.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li><li><span><a href=\"#n500r1\" data-toc-modified-id=\"n500r1-6.7\"><span class=\"toc-item-num\">6.7&nbsp;&nbsp;</span>n500r1</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-6.7.1\"><span class=\"toc-item-num\">6.7.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-6.7.2\"><span class=\"toc-item-num\">6.7.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li><li><span><a href=\"#n500r2\" data-toc-modified-id=\"n500r2-6.8\"><span class=\"toc-item-num\">6.8&nbsp;&nbsp;</span>n500r2</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-6.8.1\"><span class=\"toc-item-num\">6.8.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-6.8.2\"><span class=\"toc-item-num\">6.8.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li><li><span><a href=\"#n1000r1\" data-toc-modified-id=\"n1000r1-6.9\"><span class=\"toc-item-num\">6.9&nbsp;&nbsp;</span>n1000r1</a></span><ul class=\"toc-item\"><li><span><a href=\"#Config\" data-toc-modified-id=\"Config-6.9.1\"><span class=\"toc-item-num\">6.9.1&nbsp;&nbsp;</span>Config</a></span></li><li><span><a href=\"#Run\" data-toc-modified-id=\"Run-6.9.2\"><span class=\"toc-item-num\">6.9.2&nbsp;&nbsp;</span>Run</a></span></li></ul></li></ul></li><li><span><a href=\"#Summary\" data-toc-modified-id=\"Summary-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>Summary</a></span><ul class=\"toc-item\"><li><span><a href=\"#DB-create\" data-toc-modified-id=\"DB-create-7.1\"><span class=\"toc-item-num\">7.1&nbsp;&nbsp;</span>DB create</a></span></li><li><span><a href=\"#DB-update\" data-toc-modified-id=\"DB-update-7.2\"><span class=\"toc-item-num\">7.2&nbsp;&nbsp;</span>DB update</a></span></li></ul></li><li><span><a href=\"#sessionInfo\" data-toc-modified-id=\"sessionInfo-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>sessionInfo</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goal\n",
    "\n",
    "* Benchmark resources required to update a database\n",
    "  * Comparing Struo vs Struo2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T07:40:12.028961Z",
     "start_time": "2021-02-07T07:40:10.701Z"
    }
   },
   "outputs": [],
   "source": [
    "# working dir\n",
    "work_dir = '/ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/'\n",
    "\n",
    "# GTDB struo2 genome metadata\n",
    "metadata_file = '/ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo//metadata_1per-GTDB-Spec_gte50comp-lt5cont_wtaxID_wPath.tsv'\n",
    "\n",
    "# params\n",
    "my_seed = 831\n",
    "threads = 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T07:40:13.020441Z",
     "start_time": "2021-02-07T07:40:10.923Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘tidytable’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    dt\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(ggplot2)\n",
    "library(data.table)\n",
    "library(tidytable)\n",
    "library(LeyLabRMisc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T07:40:13.210603Z",
     "start_time": "2021-02-07T07:40:11.612Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory already exists: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/ \n"
     ]
    }
   ],
   "source": [
    "set.seed(6471)\n",
    "setDTthreads(threads)\n",
    "make_dir(work_dir)\n",
    "df.dims()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Selecting genome reps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-18T19:53:26.014425Z",
     "start_time": "2021-01-18T19:53:12.591Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 30989 × 113</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ncbi_organism_name</th><th scope=col>accession</th><th scope=col>ambiguous_bases</th><th scope=col>checkm_completeness</th><th scope=col>checkm_contamination</th><th scope=col>checkm_marker_count</th><th scope=col>checkm_marker_lineage</th><th scope=col>checkm_marker_set_count</th><th scope=col>checkm_strain_heterogeneity</th><th scope=col>coding_bases</th><th scope=col>⋯</th><th scope=col>ssu_silva_blast_perc_identity</th><th scope=col>ssu_silva_blast_subject_id</th><th scope=col>ssu_silva_taxonomy</th><th scope=col>total_gap_length</th><th scope=col>trna_aa_count</th><th scope=col>trna_count</th><th scope=col>trna_selenocysteine_count</th><th scope=col>SPECIES</th><th scope=col>gtdb_taxid</th><th scope=col>fasta_file_path</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GB_GCA_002789635.1_Candidatus Aenigmarchaeota archaeon CG_4_9_14_3_um_filter_37_18</td><td>GB_GCA_002789635.1</td><td>  9</td><td>79.44</td><td>0</td><td>149</td><td>k__Archaea (UID2)</td><td>107</td><td>0</td><td>794970</td><td>⋯</td><td>100</td><td>MNUF01000028.16163.17137</td><td>Archaea;Nanoarchaeaeota;Nanohaloarchaeia;Aenigmarchaeales;Candidatus Aenigmarchaeota archaeon CG1_02_38_14</td><td> 368</td><td>19</td><td>35</td><td>0</td><td>CG10238-14 sp002789635  </td><td> 689</td><td>/ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/archaea/GCA_002789635.1/GCA_002789635.1_ASM278963v1_genomic.fna.gz</td></tr>\n",
       "\t<tr><td>GB_GCA_001784635.1_Candidatus Micrarchaeota archaeon RBG_16_49_10                 </td><td>GB_GCA_001784635.1</td><td>151</td><td>74.92</td><td>0</td><td>149</td><td>k__Archaea (UID2)</td><td>107</td><td>0</td><td>773516</td><td>⋯</td><td>100</td><td>MFRS01000021.2138.3463  </td><td>Archaea;Nanoarchaeaeota;Nanohaloarchaeia;Aenigmarchaeales;Candidatus Micrarchaeota archaeon RBG_16_49_10  </td><td>6621</td><td>18</td><td>36</td><td>0</td><td>RBG-16-49-10 sp001784635</td><td>3509</td><td>/ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/archaea/GCA_001784635.1/GCA_001784635.1_ASM178463v1_genomic.fna.gz</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>RS_GCF_001644955.1_Pseudomonas putida</td><td>RS_GCF_001644955.1</td><td>0</td><td>99.59</td><td>0.16</td><td>813</td><td>o__Pseudomonadales (UID4488) </td><td>308</td><td>0</td><td>5508282</td><td>⋯</td><td>none</td><td>none           </td><td>none                                                                                                  </td><td>0</td><td>19</td><td>61</td><td>0</td><td>Pseudomonas_E putida_C</td><td>188090</td><td>/ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/bacteria/GCA_001644955.1/GCA_001644955.1_CBF10-2_genomic.fna.gz    </td></tr>\n",
       "\t<tr><td>RS_GCF_003233695.1_Lysobacter maris  </td><td>RS_GCF_003233695.1</td><td>0</td><td>99.77</td><td>0.69</td><td>659</td><td>f__Xanthomonadaceae (UID4214)</td><td>290</td><td>0</td><td>3531075</td><td>⋯</td><td>100 </td><td>KX886800.1.1510</td><td>Bacteria;Proteobacteria;Gammaproteobacteria;Xanthomonadales;Xanthomonadaceae;Lysobacter;Lysobacter sp.</td><td>0</td><td>20</td><td>56</td><td>0</td><td>Lysobacter maris      </td><td>183442</td><td>/ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/bacteria/GCA_003233695.1/GCA_003233695.1_ASM323369v1_genomic.fna.gz</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 30989 × 113\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " ncbi\\_organism\\_name & accession & ambiguous\\_bases & checkm\\_completeness & checkm\\_contamination & checkm\\_marker\\_count & checkm\\_marker\\_lineage & checkm\\_marker\\_set\\_count & checkm\\_strain\\_heterogeneity & coding\\_bases & ⋯ & ssu\\_silva\\_blast\\_perc\\_identity & ssu\\_silva\\_blast\\_subject\\_id & ssu\\_silva\\_taxonomy & total\\_gap\\_length & trna\\_aa\\_count & trna\\_count & trna\\_selenocysteine\\_count & SPECIES & gtdb\\_taxid & fasta\\_file\\_path\\\\\n",
       " <fct> & <fct> & <int> & <dbl> & <dbl> & <int> & <fct> & <int> & <dbl> & <int> & ⋯ & <fct> & <fct> & <fct> & <int> & <int> & <int> & <int> & <fct> & <int> & <fct>\\\\\n",
       "\\hline\n",
       "\t GB\\_GCA\\_002789635.1\\_Candidatus Aenigmarchaeota archaeon CG\\_4\\_9\\_14\\_3\\_um\\_filter\\_37\\_18 & GB\\_GCA\\_002789635.1 &   9 & 79.44 & 0 & 149 & k\\_\\_Archaea (UID2) & 107 & 0 & 794970 & ⋯ & 100 & MNUF01000028.16163.17137 & Archaea;Nanoarchaeaeota;Nanohaloarchaeia;Aenigmarchaeales;Candidatus Aenigmarchaeota archaeon CG1\\_02\\_38\\_14 &  368 & 19 & 35 & 0 & CG10238-14 sp002789635   &  689 & /ebio/abt3\\_projects/databases\\_no-backup/GTDB/release95/Struo/genomes/genbank/archaea/GCA\\_002789635.1/GCA\\_002789635.1\\_ASM278963v1\\_genomic.fna.gz\\\\\n",
       "\t GB\\_GCA\\_001784635.1\\_Candidatus Micrarchaeota archaeon RBG\\_16\\_49\\_10                  & GB\\_GCA\\_001784635.1 & 151 & 74.92 & 0 & 149 & k\\_\\_Archaea (UID2) & 107 & 0 & 773516 & ⋯ & 100 & MFRS01000021.2138.3463   & Archaea;Nanoarchaeaeota;Nanohaloarchaeia;Aenigmarchaeales;Candidatus Micrarchaeota archaeon RBG\\_16\\_49\\_10   & 6621 & 18 & 36 & 0 & RBG-16-49-10 sp001784635 & 3509 & /ebio/abt3\\_projects/databases\\_no-backup/GTDB/release95/Struo/genomes/genbank/archaea/GCA\\_001784635.1/GCA\\_001784635.1\\_ASM178463v1\\_genomic.fna.gz\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t RS\\_GCF\\_001644955.1\\_Pseudomonas putida & RS\\_GCF\\_001644955.1 & 0 & 99.59 & 0.16 & 813 & o\\_\\_Pseudomonadales (UID4488)  & 308 & 0 & 5508282 & ⋯ & none & none            & none                                                                                                   & 0 & 19 & 61 & 0 & Pseudomonas\\_E putida\\_C & 188090 & /ebio/abt3\\_projects/databases\\_no-backup/GTDB/release95/Struo/genomes/genbank/bacteria/GCA\\_001644955.1/GCA\\_001644955.1\\_CBF10-2\\_genomic.fna.gz    \\\\\n",
       "\t RS\\_GCF\\_003233695.1\\_Lysobacter maris   & RS\\_GCF\\_003233695.1 & 0 & 99.77 & 0.69 & 659 & f\\_\\_Xanthomonadaceae (UID4214) & 290 & 0 & 3531075 & ⋯ & 100  & KX886800.1.1510 & Bacteria;Proteobacteria;Gammaproteobacteria;Xanthomonadales;Xanthomonadaceae;Lysobacter;Lysobacter sp. & 0 & 20 & 56 & 0 & Lysobacter maris       & 183442 & /ebio/abt3\\_projects/databases\\_no-backup/GTDB/release95/Struo/genomes/genbank/bacteria/GCA\\_003233695.1/GCA\\_003233695.1\\_ASM323369v1\\_genomic.fna.gz\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 30989 × 113\n",
       "\n",
       "| ncbi_organism_name &lt;fct&gt; | accession &lt;fct&gt; | ambiguous_bases &lt;int&gt; | checkm_completeness &lt;dbl&gt; | checkm_contamination &lt;dbl&gt; | checkm_marker_count &lt;int&gt; | checkm_marker_lineage &lt;fct&gt; | checkm_marker_set_count &lt;int&gt; | checkm_strain_heterogeneity &lt;dbl&gt; | coding_bases &lt;int&gt; | ⋯ ⋯ | ssu_silva_blast_perc_identity &lt;fct&gt; | ssu_silva_blast_subject_id &lt;fct&gt; | ssu_silva_taxonomy &lt;fct&gt; | total_gap_length &lt;int&gt; | trna_aa_count &lt;int&gt; | trna_count &lt;int&gt; | trna_selenocysteine_count &lt;int&gt; | SPECIES &lt;fct&gt; | gtdb_taxid &lt;int&gt; | fasta_file_path &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| GB_GCA_002789635.1_Candidatus Aenigmarchaeota archaeon CG_4_9_14_3_um_filter_37_18 | GB_GCA_002789635.1 |   9 | 79.44 | 0 | 149 | k__Archaea (UID2) | 107 | 0 | 794970 | ⋯ | 100 | MNUF01000028.16163.17137 | Archaea;Nanoarchaeaeota;Nanohaloarchaeia;Aenigmarchaeales;Candidatus Aenigmarchaeota archaeon CG1_02_38_14 |  368 | 19 | 35 | 0 | CG10238-14 sp002789635   |  689 | /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/archaea/GCA_002789635.1/GCA_002789635.1_ASM278963v1_genomic.fna.gz |\n",
       "| GB_GCA_001784635.1_Candidatus Micrarchaeota archaeon RBG_16_49_10                  | GB_GCA_001784635.1 | 151 | 74.92 | 0 | 149 | k__Archaea (UID2) | 107 | 0 | 773516 | ⋯ | 100 | MFRS01000021.2138.3463   | Archaea;Nanoarchaeaeota;Nanohaloarchaeia;Aenigmarchaeales;Candidatus Micrarchaeota archaeon RBG_16_49_10   | 6621 | 18 | 36 | 0 | RBG-16-49-10 sp001784635 | 3509 | /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/archaea/GCA_001784635.1/GCA_001784635.1_ASM178463v1_genomic.fna.gz |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| RS_GCF_001644955.1_Pseudomonas putida | RS_GCF_001644955.1 | 0 | 99.59 | 0.16 | 813 | o__Pseudomonadales (UID4488)  | 308 | 0 | 5508282 | ⋯ | none | none            | none                                                                                                   | 0 | 19 | 61 | 0 | Pseudomonas_E putida_C | 188090 | /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/bacteria/GCA_001644955.1/GCA_001644955.1_CBF10-2_genomic.fna.gz     |\n",
       "| RS_GCF_003233695.1_Lysobacter maris   | RS_GCF_003233695.1 | 0 | 99.77 | 0.69 | 659 | f__Xanthomonadaceae (UID4214) | 290 | 0 | 3531075 | ⋯ | 100  | KX886800.1.1510 | Bacteria;Proteobacteria;Gammaproteobacteria;Xanthomonadales;Xanthomonadaceae;Lysobacter;Lysobacter sp. | 0 | 20 | 56 | 0 | Lysobacter maris       | 183442 | /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/bacteria/GCA_003233695.1/GCA_003233695.1_ASM323369v1_genomic.fna.gz |\n",
       "\n"
      ],
      "text/plain": [
       "      ncbi_organism_name                                                                \n",
       "1     GB_GCA_002789635.1_Candidatus Aenigmarchaeota archaeon CG_4_9_14_3_um_filter_37_18\n",
       "2     GB_GCA_001784635.1_Candidatus Micrarchaeota archaeon RBG_16_49_10                 \n",
       "⋮     ⋮                                                                                 \n",
       "30988 RS_GCF_001644955.1_Pseudomonas putida                                             \n",
       "30989 RS_GCF_003233695.1_Lysobacter maris                                               \n",
       "      accession          ambiguous_bases checkm_completeness\n",
       "1     GB_GCA_002789635.1   9             79.44              \n",
       "2     GB_GCA_001784635.1 151             74.92              \n",
       "⋮     ⋮                  ⋮               ⋮                  \n",
       "30988 RS_GCF_001644955.1 0               99.59              \n",
       "30989 RS_GCF_003233695.1 0               99.77              \n",
       "      checkm_contamination checkm_marker_count checkm_marker_lineage        \n",
       "1     0                    149                 k__Archaea (UID2)            \n",
       "2     0                    149                 k__Archaea (UID2)            \n",
       "⋮     ⋮                    ⋮                   ⋮                            \n",
       "30988 0.16                 813                 o__Pseudomonadales (UID4488) \n",
       "30989 0.69                 659                 f__Xanthomonadaceae (UID4214)\n",
       "      checkm_marker_set_count checkm_strain_heterogeneity coding_bases ⋯\n",
       "1     107                     0                           794970       ⋯\n",
       "2     107                     0                           773516       ⋯\n",
       "⋮     ⋮                       ⋮                           ⋮            ⋱\n",
       "30988 308                     0                           5508282      ⋯\n",
       "30989 290                     0                           3531075      ⋯\n",
       "      ssu_silva_blast_perc_identity ssu_silva_blast_subject_id\n",
       "1     100                           MNUF01000028.16163.17137  \n",
       "2     100                           MFRS01000021.2138.3463    \n",
       "⋮     ⋮                             ⋮                         \n",
       "30988 none                          none                      \n",
       "30989 100                           KX886800.1.1510           \n",
       "      ssu_silva_taxonomy                                                                                        \n",
       "1     Archaea;Nanoarchaeaeota;Nanohaloarchaeia;Aenigmarchaeales;Candidatus Aenigmarchaeota archaeon CG1_02_38_14\n",
       "2     Archaea;Nanoarchaeaeota;Nanohaloarchaeia;Aenigmarchaeales;Candidatus Micrarchaeota archaeon RBG_16_49_10  \n",
       "⋮     ⋮                                                                                                         \n",
       "30988 none                                                                                                      \n",
       "30989 Bacteria;Proteobacteria;Gammaproteobacteria;Xanthomonadales;Xanthomonadaceae;Lysobacter;Lysobacter sp.    \n",
       "      total_gap_length trna_aa_count trna_count trna_selenocysteine_count\n",
       "1      368             19            35         0                        \n",
       "2     6621             18            36         0                        \n",
       "⋮     ⋮                ⋮             ⋮          ⋮                        \n",
       "30988 0                19            61         0                        \n",
       "30989 0                20            56         0                        \n",
       "      SPECIES                  gtdb_taxid\n",
       "1     CG10238-14 sp002789635    689      \n",
       "2     RBG-16-49-10 sp001784635 3509      \n",
       "⋮     ⋮                        ⋮         \n",
       "30988 Pseudomonas_E putida_C   188090    \n",
       "30989 Lysobacter maris         183442    \n",
       "      fasta_file_path                                                                                                                                 \n",
       "1     /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/archaea/GCA_002789635.1/GCA_002789635.1_ASM278963v1_genomic.fna.gz \n",
       "2     /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/archaea/GCA_001784635.1/GCA_001784635.1_ASM178463v1_genomic.fna.gz \n",
       "⋮     ⋮                                                                                                                                               \n",
       "30988 /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/bacteria/GCA_001644955.1/GCA_001644955.1_CBF10-2_genomic.fna.gz    \n",
       "30989 /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/genomes/genbank/bacteria/GCA_003233695.1/GCA_003233695.1_ASM323369v1_genomic.fna.gz"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# loading metadata\n",
    "meta = read.delim(metadata_file, sep='\\t') \n",
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-18T19:53:26.016406Z",
     "start_time": "2021-01-18T19:53:17.092Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "30989"
      ],
      "text/latex": [
       "30989"
      ],
      "text/markdown": [
       "30989"
      ],
      "text/plain": [
       "[1] 30989"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# selecting dataset subsets \n",
    "pool = sample(1:nrow(meta), nrow(meta))\n",
    "pool %>% length"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Writing tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-24T07:31:03.566535Z",
     "start_time": "2021-01-24T07:31:03.515Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory already exists: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata \n"
     ]
    }
   ],
   "source": [
    "D = file.path(work_dir, 'metadata')\n",
    "make_dir(D)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### N100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:03:38.254552Z",
     "start_time": "2021-01-10T11:03:38.198Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n100_r0.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n100_r0.tsv')\n",
    "meta[pool[1:100],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:03:43.406191Z",
     "start_time": "2021-01-10T11:03:43.286Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n100_r1.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n100_r1.tsv')\n",
    "meta[pool[101:200],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:03:53.530799Z",
     "start_time": "2021-01-10T11:03:53.468Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n100_r2.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n100_r2.tsv')\n",
    "meta[pool[201:300],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:03:56.451861Z",
     "start_time": "2021-01-10T11:03:56.388Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n100_r3.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n100_r3.tsv')\n",
    "meta[pool[301:400],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### N500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:04:07.800757Z",
     "start_time": "2021-01-10T11:04:07.717Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n500_r0.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n500_r0.tsv')\n",
    "meta[pool[1001:1500],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:04:10.873404Z",
     "start_time": "2021-01-10T11:04:10.781Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n500_r1.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n500_r1.tsv')\n",
    "meta[pool[1501:2000],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:04:14.403971Z",
     "start_time": "2021-01-10T11:04:14.317Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n500_r2.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n500_r2.tsv')\n",
    "meta[pool[2001:2500],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:04:17.675727Z",
     "start_time": "2021-01-10T11:04:17.589Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n500_r3.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n500_r3.tsv')\n",
    "meta[pool[2501:3000],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### N1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:04:33.540676Z",
     "start_time": "2021-01-10T11:04:33.413Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n1000_r0.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n1000_r0.tsv')\n",
    "meta[pool[9001:10000],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:04:38.908305Z",
     "start_time": "2021-01-10T11:04:38.741Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n1000_r1.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n1000_r1.tsv')\n",
    "meta[pool[10001:11000],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:04:41.912897Z",
     "start_time": "2021-01-10T11:04:41.788Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n1000_r2.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n1000_r2.tsv')\n",
    "meta[pool[11001:12000],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T11:04:44.750590Z",
     "start_time": "2021-01-10T11:04:44.620Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n1000_r3.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n1000_r3.tsv')\n",
    "meta[pool[12001:13000],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### N10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-24T07:31:08.352077Z",
     "start_time": "2021-01-24T07:31:07.156Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n10_r1.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n10_r1.tsv')\n",
    "meta[pool[13001:13010],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-24T07:31:19.816434Z",
     "start_time": "2021-01-24T07:31:19.745Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n10_r2.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n10_r2.tsv')\n",
    "meta[pool[13011:13020],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### N250"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-24T11:47:41.537840Z",
     "start_time": "2021-01-24T11:47:41.456Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n250_r1.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n250_r1.tsv')\n",
    "meta[pool[14001:14250],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### N350"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-25T13:16:03.996931Z",
     "start_time": "2021-01-25T13:16:03.886Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n350_r1.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n350_r1.tsv')\n",
    "meta[pool[15001:15350],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### N175"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-27T10:45:53.896468Z",
     "start_time": "2021-01-27T10:45:53.779Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n175_r1.tsv \n"
     ]
    }
   ],
   "source": [
    "# writing\n",
    "outfile = file.path(D, 'metadata_n175_r1.tsv')\n",
    "meta[pool[16001:16175],] %>%\n",
    "    write_table(outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DB-create"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## N1000r0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-10T12:01:52.234666Z",
     "start_time": "2021-01-10T12:01:52.164Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- I/O --#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n1000_r0.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- databases to create --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "# Name of UniRef clustering (uniref90 or uniref50)\n",
      "## uniref90 highly recommended\n",
      "uniref_name: uniref90\n",
      "# Name of the humann3 diamond database to be created\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## This is skipped if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "#-- if custom NCBI taxdump files (or just Skip) --#\n",
      "# Used for kraken taxonomy & metaphlan\n",
      "names_dmp: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/taxdump/names.dmp\n",
      "nodes_dmp: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/taxdump/nodes.dmp\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster: --min-seq-id 0.9 -c 0.8\n",
      "    mmseqs_cluster_method: linclust         # or \"cluster\", which is slower\n",
      "  humann3:\n",
      "    batches: 100\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 --num-iterations 2 --start-sens 1 --sens-steps 3 -s 6\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    pkl: /ebio/abt3_projects2/databases_no-backup/metaphlan3/201901/mpa_v30_CHOCOPhlAn_201901.pkl\n",
      "    fasta: /ebio/abt3_projects2/databases_no-backup/metaphlan3/201901/mpa_v30_CHOCOPhlAn_201901.fna.bz2\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-create\n",
      "  config: create"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'config_db-create.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run\n",
    "\n",
    "```\n",
    "(snakemake) @ rick:/ebio/abt3_projects2/databases_no-backup/bin/struo2_1\n",
    "$ screen -L -S struo2-bmk-n1k ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/config_db-create.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DB-update"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## n10r1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-24T07:37:35.099003Z",
     "start_time": "2021-01-24T07:37:35.019Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- databases to update --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "#-- Input --#\n",
      "#--- If just a set of gene sequences to add ---#\n",
      "# If you have nucleotide/amino-acid gene sequences formatted for humann\n",
      "# If translate = True, missing nuc or AA seqs will be (rev)translated from the other, else seqs not used\n",
      "new_genes:\n",
      "  amino_acid: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.faa.gz\n",
      "  nucleotide: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.fna.gz\n",
      "  metadata: tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.txt.gz\n",
      "  translate: False\n",
      "\n",
      "#--- If a set of genomes to add ---#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/metadata/metadata_n10_r1.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# Saved databases that will be updated\n",
      "kraken2_db:\n",
      "  library:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/library/\n",
      "  taxonomy: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/taxonomy/\n",
      "genes_db:\n",
      "  genes:\n",
      "    mmseqs_db:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genes_db.tar.gz\n",
      "    amino_acid: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.faa.gz\n",
      "    nucleotide: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.fna.gz\n",
      "    metadata:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.txt.gz\n",
      "  cluster:\n",
      "    mmseqs_db:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_db.tar.gz    \n",
      "humann_db:\n",
      "  query:\n",
      "    hits: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/humann3/annotation_hits.gz\n",
      "  cluster:\n",
      "    reps: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_reps.faa.gz\n",
      "    membership: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_membership.tsv.gz\n",
      "\n",
      "#-- Output --#\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n10r1/\n",
      "\n",
      "# Name of the humann3 diamond database to create\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## Skip if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster_update: --min-seq-id 0.9 -c 0.8 -s 4.0    \n",
      "  humann3:\n",
      "    batches: 10\n",
      "    filter_existing: --min-pident 0  # any existing genes w/ < cutoff with be re-queried\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 -s 6 --num-iterations 2\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-update\n",
      "  config: update"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'n10r1', 'config-update.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run\n",
    "\n",
    "```\n",
    "screen -L -S struo2-update ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n10r1/config-update.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## n100r1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-12T13:52:23.857253Z",
     "start_time": "2021-01-12T13:52:23.764Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- databases to update --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "#-- Input --#\n",
      "#--- If just a set of gene sequences to add ---#\n",
      "# If you have nucleotide/amino-acid gene sequences formatted for humann\n",
      "# If translate = True, missing nuc or AA seqs will be (rev)translated from the other, else seqs not used\n",
      "new_genes:\n",
      "  amino_acid: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.faa.gz\n",
      "  nucleotide: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.fna.gz\n",
      "  metadata: tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.txt.gz\n",
      "  translate: False\n",
      "\n",
      "#--- If a set of genomes to add ---#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90//metadata/metadata_n100_r1.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# Saved databases that will be updated\n",
      "kraken2_db:\n",
      "  library:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/library/\n",
      "  taxonomy: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/taxonomy/\n",
      "genes_db:\n",
      "  genes:\n",
      "    mmseqs_db:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genes_db.tar.gz\n",
      "    amino_acid: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.faa.gz\n",
      "    nucleotide: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.fna.gz\n",
      "    metadata:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.txt.gz\n",
      "  cluster:\n",
      "    mmseqs_db:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_db.tar.gz    \n",
      "humann_db:\n",
      "  query:\n",
      "    hits: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/humann3/annotation_hits.gz\n",
      "  cluster:\n",
      "    reps: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_reps.faa.gz\n",
      "    membership: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_membership.tsv.gz\n",
      "\n",
      "#-- Output --#\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n100r1/\n",
      "\n",
      "# Name of the humann3 diamond database to create\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## Skip if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster_update: --min-seq-id 0.9 -c 0.8 -s 4.0    \n",
      "  humann3:\n",
      "    batches: 10\n",
      "    filter_existing: --min-pident 0  # any existing genes w/ < cutoff with be re-queried\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 -s 6 --num-iterations 2\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-update\n",
      "  config: update"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'n100r1', 'config-update.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "screen -L -S struo2-update ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n100r1/config-update.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## n100r2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T07:57:38.549612Z",
     "start_time": "2021-01-21T07:57:38.162Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- databases to update --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "#-- Input --#\n",
      "#--- If just a set of gene sequences to add ---#\n",
      "# If you have nucleotide/amino-acid gene sequences formatted for humann\n",
      "# If translate = True, missing nuc or AA seqs will be (rev)translated from the other, else seqs not used\n",
      "new_genes:\n",
      "  amino_acid: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.faa.gz\n",
      "  nucleotide: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.fna.gz\n",
      "  metadata: tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.txt.gz\n",
      "  translate: False\n",
      "\n",
      "#--- If a set of genomes to add ---#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/metadata/metadata_n100_r2.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# Saved databases that will be updated\n",
      "kraken2_db:\n",
      "  library:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/library/\n",
      "  taxonomy: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/taxonomy/\n",
      "genes_db:\n",
      "  genes:\n",
      "    mmseqs_db:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genes_db.tar.gz\n",
      "    amino_acid: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.faa.gz\n",
      "    nucleotide: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.fna.gz\n",
      "    metadata:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.txt.gz\n",
      "  cluster:\n",
      "    mmseqs_db:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_db.tar.gz    \n",
      "humann_db:\n",
      "  query:\n",
      "    hits: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/humann3/annotation_hits.gz\n",
      "  cluster:\n",
      "    reps: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_reps.faa.gz\n",
      "    membership: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_membership.tsv.gz\n",
      "\n",
      "#-- Output --#\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n100r2/\n",
      "\n",
      "# Name of the humann3 diamond database to create\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## Skip if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster_update: --min-seq-id 0.9 -c 0.8 -s 4.0    \n",
      "  humann3:\n",
      "    batches: 10\n",
      "    filter_existing: --min-pident 0  # any existing genes w/ < cutoff with be re-queried\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 -s 6 --num-iterations 2\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-update\n",
      "  config: update"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'n100r2', 'config-update.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "(snakemake) @ rick:/ebio/abt3_projects2/databases_no-backup/bin/struo2_1\n",
    "$ screen -L -S struo2-n100r2 ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n100r2/config-update.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## n175r1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-27T10:49:34.898591Z",
     "start_time": "2021-01-27T10:49:34.770Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- databases to update --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "#-- Input --#\n",
      "#--- If just a set of gene sequences to add ---#\n",
      "# If you have nucleotide/amino-acid gene sequences formatted for humann\n",
      "# If translate = True, missing nuc or AA seqs will be (rev)translated from the other, else seqs not used\n",
      "new_genes:\n",
      "  amino_acid: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.faa.gz\n",
      "  nucleotide: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.fna.gz\n",
      "  metadata: tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.txt.gz\n",
      "  translate: False\n",
      "\n",
      "#--- If a set of genomes to add ---#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/metadata/metadata_n175_r1.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# Saved databases that will be updated\n",
      "kraken2_db:\n",
      "  library:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/library/\n",
      "  taxonomy: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/taxonomy/\n",
      "genes_db:\n",
      "  genes:\n",
      "    mmseqs_db:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genes_db.tar.gz\n",
      "    amino_acid: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.faa.gz\n",
      "    nucleotide: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.fna.gz\n",
      "    metadata:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.txt.gz\n",
      "  cluster:\n",
      "    mmseqs_db:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_db.tar.gz    \n",
      "humann_db:\n",
      "  query:\n",
      "    hits: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/humann3/annotation_hits.gz\n",
      "  cluster:\n",
      "    reps: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_reps.faa.gz\n",
      "    membership: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_membership.tsv.gz\n",
      "\n",
      "#-- Output --#\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n175r1/\n",
      "\n",
      "# Name of the humann3 diamond database to create\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## Skip if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster_update: --min-seq-id 0.9 -c 0.8 -s 4.0    \n",
      "  humann3:\n",
      "    batches: 10\n",
      "    filter_existing: --min-pident 0  # any existing genes w/ < cutoff with be re-queried\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 -s 6 --num-iterations 2\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-update\n",
      "  config: update"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'n175r1', 'config-update.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "(snakemake) @ rick:/ebio/abt3_projects2/databases_no-backup/bin/struo2_1\n",
    "$ screen -L -S struo2-n175r1 ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n175r1/config-update.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## n250r1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-24T11:49:16.152047Z",
     "start_time": "2021-01-24T11:49:16.094Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- databases to update --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "#-- Input --#\n",
      "#--- If just a set of gene sequences to add ---#\n",
      "# If you have nucleotide/amino-acid gene sequences formatted for humann\n",
      "# If translate = True, missing nuc or AA seqs will be (rev)translated from the other, else seqs not used\n",
      "new_genes:\n",
      "  amino_acid: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.faa.gz\n",
      "  nucleotide: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.fna.gz\n",
      "  metadata: tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.txt.gz\n",
      "  translate: False\n",
      "\n",
      "#--- If a set of genomes to add ---#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/metadata/metadata_n250_r1.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# Saved databases that will be updated\n",
      "kraken2_db:\n",
      "  library:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/library/\n",
      "  taxonomy: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/taxonomy/\n",
      "genes_db:\n",
      "  genes:\n",
      "    mmseqs_db:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genes_db.tar.gz\n",
      "    amino_acid: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.faa.gz\n",
      "    nucleotide: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.fna.gz\n",
      "    metadata:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.txt.gz\n",
      "  cluster:\n",
      "    mmseqs_db:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_db.tar.gz    \n",
      "humann_db:\n",
      "  query:\n",
      "    hits: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/humann3/annotation_hits.gz\n",
      "  cluster:\n",
      "    reps: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_reps.faa.gz\n",
      "    membership: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_membership.tsv.gz\n",
      "\n",
      "#-- Output --#\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n250r1/\n",
      "\n",
      "# Name of the humann3 diamond database to create\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## Skip if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster_update: --min-seq-id 0.9 -c 0.8 -s 4.0    \n",
      "  humann3:\n",
      "    batches: 10\n",
      "    filter_existing: --min-pident 0  # any existing genes w/ < cutoff with be re-queried\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 -s 6 --num-iterations 2\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-update\n",
      "  config: update"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'n250r1', 'config-update.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "(snakemake) @ rick:/ebio/abt3_projects2/databases_no-backup/bin/struo2_1\n",
    "$ screen -L -S struo2-n250r1 ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n250r1/config-update.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## n350r1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-25T13:25:43.339072Z",
     "start_time": "2021-01-25T13:25:42.579Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- databases to update --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "#-- Input --#\n",
      "#--- If just a set of gene sequences to add ---#\n",
      "# If you have nucleotide/amino-acid gene sequences formatted for humann\n",
      "# If translate = True, missing nuc or AA seqs will be (rev)translated from the other, else seqs not used\n",
      "new_genes:\n",
      "  amino_acid: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.faa.gz\n",
      "  nucleotide: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.fna.gz\n",
      "  metadata: tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.txt.gz\n",
      "  translate: False\n",
      "\n",
      "#--- If a set of genomes to add ---#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/metadata/metadata_n350_r1.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# Saved databases that will be updated\n",
      "kraken2_db:\n",
      "  library:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/library/\n",
      "  taxonomy: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/taxonomy/\n",
      "genes_db:\n",
      "  genes:\n",
      "    mmseqs_db:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genes_db.tar.gz\n",
      "    amino_acid: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.faa.gz\n",
      "    nucleotide: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.fna.gz\n",
      "    metadata:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.txt.gz\n",
      "  cluster:\n",
      "    mmseqs_db:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_db.tar.gz    \n",
      "humann_db:\n",
      "  query:\n",
      "    hits: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/humann3/annotation_hits.gz\n",
      "  cluster:\n",
      "    reps: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_reps.faa.gz\n",
      "    membership: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_membership.tsv.gz\n",
      "\n",
      "#-- Output --#\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n350r1/\n",
      "\n",
      "# Name of the humann3 diamond database to create\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## Skip if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster_update: --min-seq-id 0.9 -c 0.8 -s 4.0    \n",
      "  humann3:\n",
      "    batches: 10\n",
      "    filter_existing: --min-pident 0  # any existing genes w/ < cutoff with be re-queried\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 -s 6 --num-iterations 2\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-update\n",
      "  config: update"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'n350r1', 'config-update.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "$ screen -L -S struo2-n350r1 ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n350r1/config-update.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## n500r1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-18T20:09:47.270268Z",
     "start_time": "2021-01-18T20:09:47.151Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- databases to update --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "#-- Input --#\n",
      "#--- If just a set of gene sequences to add ---#\n",
      "# If you have nucleotide/amino-acid gene sequences formatted for humann\n",
      "# If translate = True, missing nuc or AA seqs will be (rev)translated from the other, else seqs not used\n",
      "new_genes:\n",
      "  amino_acid: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.faa.gz\n",
      "  nucleotide: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.fna.gz\n",
      "  metadata: tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.txt.gz\n",
      "  translate: False\n",
      "\n",
      "#--- If a set of genomes to add ---#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/metadata/metadata_n500_r1.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# Saved databases that will be updated\n",
      "kraken2_db:\n",
      "  library:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/library/\n",
      "  taxonomy: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/taxonomy/\n",
      "genes_db:\n",
      "  genes:\n",
      "    mmseqs_db:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genes_db.tar.gz\n",
      "    amino_acid: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.faa.gz\n",
      "    nucleotide: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.fna.gz\n",
      "    metadata:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.txt.gz\n",
      "  cluster:\n",
      "    mmseqs_db:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_db.tar.gz    \n",
      "humann_db:\n",
      "  query:\n",
      "    hits: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/humann3/annotation_hits.gz\n",
      "  cluster:\n",
      "    reps: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_reps.faa.gz\n",
      "    membership: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_membership.tsv.gz\n",
      "\n",
      "#-- Output --#\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n500r1/\n",
      "\n",
      "# Name of the humann3 diamond database to create\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## Skip if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster_update: --min-seq-id 0.9 -c 0.8 -s 4.0    \n",
      "  humann3:\n",
      "    batches: 10\n",
      "    filter_existing: --min-pident 0  # any existing genes w/ < cutoff with be re-queried\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 -s 6 --num-iterations 2\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-update\n",
      "  config: update"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'n500r1', 'config-update.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "(snakemake) @ rick:/ebio/abt3_projects2/databases_no-backup/bin/struo2_1\n",
    "$ screen -L -S struo2-db-update-n500 ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n500r1/config-update.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## n500r2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T19:40:50.852643Z",
     "start_time": "2021-01-21T19:40:50.708Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- databases to update --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "#-- Input --#\n",
      "#--- If just a set of gene sequences to add ---#\n",
      "# If you have nucleotide/amino-acid gene sequences formatted for humann\n",
      "# If translate = True, missing nuc or AA seqs will be (rev)translated from the other, else seqs not used\n",
      "new_genes:\n",
      "  amino_acid: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.faa.gz\n",
      "  nucleotide: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.fna.gz\n",
      "  metadata: tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.txt.gz\n",
      "  translate: False\n",
      "\n",
      "#--- If a set of genomes to add ---#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/metadata/metadata_n500_r2.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# Saved databases that will be updated\n",
      "kraken2_db:\n",
      "  library:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/library/\n",
      "  taxonomy: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/taxonomy/\n",
      "genes_db:\n",
      "  genes:\n",
      "    mmseqs_db:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genes_db.tar.gz\n",
      "    amino_acid: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.faa.gz\n",
      "    nucleotide: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.fna.gz\n",
      "    metadata:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.txt.gz\n",
      "  cluster:\n",
      "    mmseqs_db:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_db.tar.gz    \n",
      "humann_db:\n",
      "  query:\n",
      "    hits: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/humann3/annotation_hits.gz\n",
      "  cluster:\n",
      "    reps: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_reps.faa.gz\n",
      "    membership: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_membership.tsv.gz\n",
      "\n",
      "#-- Output --#\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n500r2/\n",
      "\n",
      "# Name of the humann3 diamond database to create\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## Skip if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster_update: --min-seq-id 0.9 -c 0.8 -s 4.0    \n",
      "  humann3:\n",
      "    batches: 10\n",
      "    filter_existing: --min-pident 0  # any existing genes w/ < cutoff with be re-queried\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 -s 6 --num-iterations 2\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-update\n",
      "  config: update"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'n500r2', 'config-update.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "(snakemake) @ rick:/ebio/abt3_projects2/databases_no-backup/bin/struo2_1\n",
    "$ screen -L -S struo2-n500r2 ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n500r2/config-update.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## n1000r1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-13T07:59:21.480243Z",
     "start_time": "2021-01-13T07:59:21.418Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-- email notifications of pipeline success/failure (use \"Skip\" to deactivate) --#\n",
      "email: nyoungblut@tuebingen.mpg.de\n",
      "\n",
      "#-- databases to update --#\n",
      "# Replace \"Create\" with \"Skip\" to skip creation of any of these\n",
      "# Note that braken relies on the kraken2 database\n",
      "databases:\n",
      "  kraken2: Create\n",
      "  bracken: Create\n",
      "  genes: Create\n",
      "  humann3_bowtie2: Create\n",
      "  humann3_diamond: Create\n",
      "  metaphlan3: Skip #Create\n",
      "\n",
      "#-- Input --#\n",
      "#--- If just a set of gene sequences to add ---#\n",
      "# If you have nucleotide/amino-acid gene sequences formatted for humann\n",
      "# If translate = True, missing nuc or AA seqs will be (rev)translated from the other, else seqs not used\n",
      "new_genes:\n",
      "  amino_acid: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.faa.gz\n",
      "  nucleotide: Skip #tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.fna.gz\n",
      "  metadata: tests/output_GTDBr89_n5/genes/UniRef50/genome_reps_filtered.txt.gz\n",
      "  translate: False\n",
      "\n",
      "#--- If a set of genomes to add ---#\n",
      "# file listing samples and associated data\n",
      "samples_file: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/metadata/metadata_n1000_r1.tsv\n",
      "\n",
      "## column names in samples table\n",
      "samples_col: 'ncbi_organism_name'\n",
      "accession_col: 'accession'\n",
      "fasta_file_path_col: 'fasta_file_path'\n",
      "taxID_col: 'gtdb_taxid'          # or 'ncbi_species_taxid'\n",
      "taxonomy_col: 'gtdb_taxonomy'    # or 'ncbi_taxonomy' \n",
      "\n",
      "# Saved databases that will be updated\n",
      "kraken2_db:\n",
      "  library:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/library/\n",
      "  taxonomy: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/kraken2/taxonomy/\n",
      "genes_db:\n",
      "  genes:\n",
      "    mmseqs_db:  /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genes_db.tar.gz\n",
      "    amino_acid: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.faa.gz\n",
      "    nucleotide: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.fna.gz\n",
      "    metadata:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/genome_reps_filtered.txt.gz\n",
      "  cluster:\n",
      "    mmseqs_db:   /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_db.tar.gz    \n",
      "humann_db:\n",
      "  query:\n",
      "    hits: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/humann3/annotation_hits.gz\n",
      "  cluster:\n",
      "    reps: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_reps.faa.gz\n",
      "    membership: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/genes/cluster/clusters_membership.tsv.gz\n",
      "\n",
      "#-- Output --#\n",
      "# output location\n",
      "output_dir: /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n1000r1/\n",
      "\n",
      "# Name of the humann3 diamond database to create\n",
      "## This must match naming allowed by humann3\n",
      "dmnd_name: uniref90_201901.dmnd\n",
      "# Index mapping UniRef90 clusters to UniRef50 (saves time vs re-annotating)\n",
      "## Skip if annotating with UniRef50\n",
      "cluster_idx: /ebio/abt3_projects2/databases_no-backup/uniref/2019.01/uniref50-90.pkl\n",
      "\n",
      "# temporary file directory (your username will be added automatically)\n",
      "tmp_dir: /ebio/abt3_scratch/\n",
      "\n",
      "#-- keep intermediate files required for re-creating DBs (eg., w/ more genomes) --#\n",
      "# If \"True\", the intermediate files are saved to `output_dir`\n",
      "# Else, the intermediate files are temporarily stored in `temp_folder`\n",
      "keep_intermediate: True\n",
      "\n",
      "#-- software parameters --#\n",
      "# `vsearch_per_genome` = per-genome gene clustering\n",
      "# for humann3, use either mmseqs or diamond (mmseqs gets priority if neither skipped)\n",
      "# for humann3::mmseqs_search::run, --num-iterations must be >=2\n",
      "params:\n",
      "  ionice: -c 3\n",
      "  bracken:\n",
      "    build_kmer: 35\n",
      "    build_read_lens:\n",
      "      - 100\n",
      "      - 150\n",
      "  genes:\n",
      "    prodigal: \"\"\n",
      "    vsearch_per_genome: --id 0.97 --strand both --qmask none --fasta_width 0\n",
      "    mmseqs_cluster_update: --min-seq-id 0.9 -c 0.8 -s 4.0    \n",
      "  humann3:\n",
      "    batches: 10\n",
      "    filter_existing: --min-pident 0  # any existing genes w/ < cutoff with be re-queried\n",
      "    mmseqs_search:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/mmseqs/UniRef90/2019_01/uniref90\n",
      "      run: -e 1e-3 --max-accept 1 --max-seqs 100 -s 6 --num-iterations 2\n",
      "    diamond:\n",
      "      db: /ebio/abt3_projects2/databases_no-backup/humann3/201901/uniref90/uniref90_201901.dmnd\n",
      "      run: Skip #--evalue 1e-3 --query-cover 80 --id 90 --max-target-seqs 1 --block-size 4 --index-chunks 2\n",
      "    propagate_annotations: --min-cov 80 --min-pident 90\n",
      "  metaphlan3:\n",
      "    species_specific_markers: --ext-abs-cutoff 10 --ext-rel-cutoff 0.1\n",
      "\n",
      "#-- snakemake pipeline --#\n",
      "pipeline:\n",
      "  snakemake_folder: ./\n",
      "  script_folder: ./bin/scripts/\n",
      "  name: Struo2_db-update\n",
      "  config: update"
     ]
    }
   ],
   "source": [
    "config_file = file.path(work_dir, 'n1000r0', 'struo2', 'n1000r1', 'config-update.yaml')\n",
    "cat_file(config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "screen -L -S struo2-update ./snakemake_sge.sh /ebio/abt3_projects2/databases_no-backup/GTDB/release95/Struo/benchmarking/db_update/UniRef50-90/n1000r0/struo2/n100r1/config-update.yaml 50 -F\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DB create"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-18T20:04:29.450868Z",
     "start_time": "2021-01-18T20:04:29.401Z"
    }
   },
   "outputs": [],
   "source": [
    "#' reading in benchmark all files \n",
    "read_bmk = function(F){\n",
    "    df = read.delim(F, sep='\\t')\n",
    "    df$file = basename(F)\n",
    "    df$rule = basename(dirname(F))\n",
    "    return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-18T20:04:29.831719Z",
     "start_time": "2021-01-18T20:04:29.690Z"
    }
   },
   "outputs": [],
   "source": [
    "#' listing & reading in all benchmark files from a base path\n",
    "read_bmks = function(F){\n",
    "    df = list_files(F, '*.txt') %>%\n",
    "        lapply(read_bmk) %>%\n",
    "        do.call(rbind, .)\n",
    "    if(grepl('uniref[0-9]+/benchmarks', F)){\n",
    "        df$dataset = F %>% dirname %>% dirname %>% dirname %>% basename\n",
    "        df$pipeline = F %>% dirname %>% dirname %>% basename\n",
    "    } else {\n",
    "        df$dataset = F %>% dirname %>% dirname %>% basename\n",
    "        df$pipeline = F %>% dirname %>% basename\n",
    "    }\n",
    "\n",
    "    return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-18T20:04:42.894828Z",
     "start_time": "2021-01-18T20:04:29.817Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 5322 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>s</th><th scope=col>h.m.s</th><th scope=col>max_rss</th><th scope=col>max_vms</th><th scope=col>max_uss</th><th scope=col>max_pss</th><th scope=col>io_in</th><th scope=col>io_out</th><th scope=col>mean_load</th><th scope=col>cpu_time</th><th scope=col>file</th><th scope=col>rule</th><th scope=col>dataset</th><th scope=col>pipeline</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>73.9160</td><td>0:01:13</td><td>8468.47</td><td>9393.21</td><td>8455.30</td><td>8455.85</td><td>9190.66</td><td>314.50</td><td>706.48</td><td>523.02</td><td>ReadLen100.txt</td><td>bracken_build</td><td>n1000r0</td><td>struo2</td></tr>\n",
       "\t<tr><td>87.2083</td><td>0:01:27</td><td>8462.74</td><td>9587.34</td><td>8450.84</td><td>8451.15</td><td>9477.97</td><td>365.96</td><td>687.95</td><td>403.13</td><td>ReadLen150.txt</td><td>bracken_build</td><td>n1000r0</td><td>struo2</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td> 179.6957</td><td>0:02:59</td><td>1541.27</td><td>1848.79</td><td>1535.45</td><td>1535.91</td><td>2.37</td><td>0</td><td>298.75</td><td>  537.17</td><td>RS_GCF_900636745_1_Actinobacillus_equuli.txt     </td><td>vsearch_per_genome</td><td>n1000r0</td><td>struo2</td></tr>\n",
       "\t<tr><td>3396.6671</td><td>0:56:36</td><td>1579.31</td><td>1878.46</td><td>1575.07</td><td>1575.19</td><td>7.39</td><td>0</td><td>372.96</td><td>12668.65</td><td>RS_GCF_902150045_1_Chitinophaga_polysaccharea.txt</td><td>vsearch_per_genome</td><td>n1000r0</td><td>struo2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5322 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " s & h.m.s & max\\_rss & max\\_vms & max\\_uss & max\\_pss & io\\_in & io\\_out & mean\\_load & cpu\\_time & file & rule & dataset & pipeline\\\\\n",
       " <dbl> & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 73.9160 & 0:01:13 & 8468.47 & 9393.21 & 8455.30 & 8455.85 & 9190.66 & 314.50 & 706.48 & 523.02 & ReadLen100.txt & bracken\\_build & n1000r0 & struo2\\\\\n",
       "\t 87.2083 & 0:01:27 & 8462.74 & 9587.34 & 8450.84 & 8451.15 & 9477.97 & 365.96 & 687.95 & 403.13 & ReadLen150.txt & bracken\\_build & n1000r0 & struo2\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t  179.6957 & 0:02:59 & 1541.27 & 1848.79 & 1535.45 & 1535.91 & 2.37 & 0 & 298.75 &   537.17 & RS\\_GCF\\_900636745\\_1\\_Actinobacillus\\_equuli.txt      & vsearch\\_per\\_genome & n1000r0 & struo2\\\\\n",
       "\t 3396.6671 & 0:56:36 & 1579.31 & 1878.46 & 1575.07 & 1575.19 & 7.39 & 0 & 372.96 & 12668.65 & RS\\_GCF\\_902150045\\_1\\_Chitinophaga\\_polysaccharea.txt & vsearch\\_per\\_genome & n1000r0 & struo2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5322 × 14\n",
       "\n",
       "| s &lt;dbl&gt; | h.m.s &lt;fct&gt; | max_rss &lt;dbl&gt; | max_vms &lt;dbl&gt; | max_uss &lt;dbl&gt; | max_pss &lt;dbl&gt; | io_in &lt;dbl&gt; | io_out &lt;dbl&gt; | mean_load &lt;dbl&gt; | cpu_time &lt;dbl&gt; | file &lt;chr&gt; | rule &lt;chr&gt; | dataset &lt;chr&gt; | pipeline &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 73.9160 | 0:01:13 | 8468.47 | 9393.21 | 8455.30 | 8455.85 | 9190.66 | 314.50 | 706.48 | 523.02 | ReadLen100.txt | bracken_build | n1000r0 | struo2 |\n",
       "| 87.2083 | 0:01:27 | 8462.74 | 9587.34 | 8450.84 | 8451.15 | 9477.97 | 365.96 | 687.95 | 403.13 | ReadLen150.txt | bracken_build | n1000r0 | struo2 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "|  179.6957 | 0:02:59 | 1541.27 | 1848.79 | 1535.45 | 1535.91 | 2.37 | 0 | 298.75 |   537.17 | RS_GCF_900636745_1_Actinobacillus_equuli.txt      | vsearch_per_genome | n1000r0 | struo2 |\n",
       "| 3396.6671 | 0:56:36 | 1579.31 | 1878.46 | 1575.07 | 1575.19 | 7.39 | 0 | 372.96 | 12668.65 | RS_GCF_902150045_1_Chitinophaga_polysaccharea.txt | vsearch_per_genome | n1000r0 | struo2 |\n",
       "\n"
      ],
      "text/plain": [
       "     s         h.m.s   max_rss max_vms max_uss max_pss io_in   io_out mean_load\n",
       "1    73.9160   0:01:13 8468.47 9393.21 8455.30 8455.85 9190.66 314.50 706.48   \n",
       "2    87.2083   0:01:27 8462.74 9587.34 8450.84 8451.15 9477.97 365.96 687.95   \n",
       "⋮    ⋮         ⋮       ⋮       ⋮       ⋮       ⋮       ⋮       ⋮      ⋮        \n",
       "5321  179.6957 0:02:59 1541.27 1848.79 1535.45 1535.91 2.37    0      298.75   \n",
       "5322 3396.6671 0:56:36 1579.31 1878.46 1575.07 1575.19 7.39    0      372.96   \n",
       "     cpu_time file                                             \n",
       "1    523.02   ReadLen100.txt                                   \n",
       "2    403.13   ReadLen150.txt                                   \n",
       "⋮    ⋮        ⋮                                                \n",
       "5321   537.17 RS_GCF_900636745_1_Actinobacillus_equuli.txt     \n",
       "5322 12668.65 RS_GCF_902150045_1_Chitinophaga_polysaccharea.txt\n",
       "     rule               dataset pipeline\n",
       "1    bracken_build      n1000r0 struo2  \n",
       "2    bracken_build      n1000r0 struo2  \n",
       "⋮    ⋮                  ⋮       ⋮       \n",
       "5321 vsearch_per_genome n1000r0 struo2  \n",
       "5322 vsearch_per_genome n1000r0 struo2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# reading in all benchmark files\n",
    "L = list(\n",
    "    file.path(work_dir, 'n1000r0', 'struo2', 'benchmarks')\n",
    ") \n",
    "bmk_create = L %>%\n",
    "    lapply(read_bmks) %>%\n",
    "    do.call(rbind, .)\n",
    "bmk_create"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-18T20:04:42.919092Z",
     "start_time": "2021-01-18T20:04:29.920Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A tibble: 1 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset</th><th scope=col>pipeline</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>5322</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1 × 3\n",
       "\\begin{tabular}{lll}\n",
       " dataset & pipeline & n\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t n1000r0 & struo2 & 5322\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1 × 3\n",
       "\n",
       "| dataset &lt;chr&gt; | pipeline &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| n1000r0 | struo2 | 5322 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset pipeline n   \n",
       "1 n1000r0 struo2   5322"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# records per dataset\n",
    "bmk_create %>%\n",
    "    group_by(dataset, pipeline) %>%\n",
    "    summarize(n = n(), .groups='drop')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-18T20:04:42.950008Z",
     "start_time": "2021-01-18T20:04:30.130Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A tibble: 1 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset</th><th scope=col>pipeline</th><th scope=col>cpu_hours</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>602.5677</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1 × 3\n",
       "\\begin{tabular}{lll}\n",
       " dataset & pipeline & cpu\\_hours\\\\\n",
       " <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t n1000r0 & struo2 & 602.5677\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1 × 3\n",
       "\n",
       "| dataset &lt;chr&gt; | pipeline &lt;chr&gt; | cpu_hours &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| n1000r0 | struo2 | 602.5677 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset pipeline cpu_hours\n",
       "1 n1000r0 struo2   602.5677 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#' cpu hours\n",
    "bmk_create_s = bmk_create %>%\n",
    "    group_by(dataset, pipeline) %>%\n",
    "    summarize(cpu_hours = sum(s) / (60 * 60),\n",
    "              .groups='drop')\n",
    "df.dims(20)\n",
    "bmk_create_s %>% arrange(dataset, pipeline)\n",
    "df.dims()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DB update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-28T11:54:11.790936Z",
     "start_time": "2021-01-28T11:54:11.761Z"
    }
   },
   "outputs": [],
   "source": [
    "#' listing & reading in all benchmark files from a base path\n",
    "read_bmks = function(F){\n",
    "    df = list_files(F, '*.txt') %>%\n",
    "        lapply(read_bmk) %>%\n",
    "        do.call(rbind, .)\n",
    "    if(grepl('uniref[0-9]+/benchmarks', F)){\n",
    "        df$dataset_start = F %>% dirname %>% dirname %>% dirname %>% dirname %>% basename\n",
    "        df$pipeline = F %>% dirname %>% dirname %>% dirname %>% basename\n",
    "        df$dataset_add = F %>% dirname %>% dirname %>% basename\n",
    "    } else {\n",
    "        df$dataset_start = F %>% dirname %>% dirname %>% dirname %>% basename\n",
    "        df$pipeline = F %>% dirname %>% dirname %>% basename\n",
    "        df$dataset_add = F %>% dirname %>% basename\n",
    "    }\n",
    "    return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-28T11:54:38.999104Z",
     "start_time": "2021-01-28T11:54:12.014Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 7243 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>s</th><th scope=col>h.m.s</th><th scope=col>max_rss</th><th scope=col>max_vms</th><th scope=col>max_uss</th><th scope=col>max_pss</th><th scope=col>io_in</th><th scope=col>io_out</th><th scope=col>mean_load</th><th scope=col>cpu_time</th><th scope=col>file</th><th scope=col>rule</th><th scope=col>dataset_start</th><th scope=col>pipeline</th><th scope=col>dataset_add</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>75.4033</td><td>0:01:15</td><td>7689   </td><td>8595.88</td><td>7677.05</td><td>7679.26</td><td>9586.42</td><td>361.28</td><td>146.61</td><td>33.20</td><td>ReadLen100.txt</td><td>bracken_build</td><td>n1000r0</td><td>struo2</td><td>n10r1</td></tr>\n",
       "\t<tr><td>42.2097</td><td>0:00:42</td><td>7569.01</td><td>8378.32</td><td>7557.48</td><td>7559.79</td><td>6503.82</td><td>43.92 </td><td>128.93</td><td>54.52</td><td>ReadLen150.txt</td><td>bracken_build</td><td>n1000r0</td><td>struo2</td><td>n10r1</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>55.2952</td><td>0:00:55</td><td>1536.54</td><td>1846.98</td><td>1531.92</td><td>1532.43</td><td>2.77</td><td>0</td><td>270.36</td><td>149.79</td><td>RS_GCF_900625105_1_Agathobaculum_sp_Marseille_P7918.txt</td><td>vsearch_per_genome</td><td>n1000r0</td><td>struo2</td><td>n500r1</td></tr>\n",
       "\t<tr><td>40.2421</td><td>0:00:40</td><td>1508.84</td><td>1820.2 </td><td>1504.31</td><td>1504.45</td><td>1.8 </td><td>0</td><td> 93.77</td><td> 38.20</td><td>RS_GCF_900638635_1_Bartonella_vinsonii.txt             </td><td>vsearch_per_genome</td><td>n1000r0</td><td>struo2</td><td>n500r1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 7243 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " s & h.m.s & max\\_rss & max\\_vms & max\\_uss & max\\_pss & io\\_in & io\\_out & mean\\_load & cpu\\_time & file & rule & dataset\\_start & pipeline & dataset\\_add\\\\\n",
       " <dbl> & <fct> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 75.4033 & 0:01:15 & 7689    & 8595.88 & 7677.05 & 7679.26 & 9586.42 & 361.28 & 146.61 & 33.20 & ReadLen100.txt & bracken\\_build & n1000r0 & struo2 & n10r1\\\\\n",
       "\t 42.2097 & 0:00:42 & 7569.01 & 8378.32 & 7557.48 & 7559.79 & 6503.82 & 43.92  & 128.93 & 54.52 & ReadLen150.txt & bracken\\_build & n1000r0 & struo2 & n10r1\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 55.2952 & 0:00:55 & 1536.54 & 1846.98 & 1531.92 & 1532.43 & 2.77 & 0 & 270.36 & 149.79 & RS\\_GCF\\_900625105\\_1\\_Agathobaculum\\_sp\\_Marseille\\_P7918.txt & vsearch\\_per\\_genome & n1000r0 & struo2 & n500r1\\\\\n",
       "\t 40.2421 & 0:00:40 & 1508.84 & 1820.2  & 1504.31 & 1504.45 & 1.8  & 0 &  93.77 &  38.20 & RS\\_GCF\\_900638635\\_1\\_Bartonella\\_vinsonii.txt              & vsearch\\_per\\_genome & n1000r0 & struo2 & n500r1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 7243 × 15\n",
       "\n",
       "| s &lt;dbl&gt; | h.m.s &lt;fct&gt; | max_rss &lt;chr&gt; | max_vms &lt;chr&gt; | max_uss &lt;chr&gt; | max_pss &lt;chr&gt; | io_in &lt;chr&gt; | io_out &lt;chr&gt; | mean_load &lt;dbl&gt; | cpu_time &lt;dbl&gt; | file &lt;chr&gt; | rule &lt;chr&gt; | dataset_start &lt;chr&gt; | pipeline &lt;chr&gt; | dataset_add &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 75.4033 | 0:01:15 | 7689    | 8595.88 | 7677.05 | 7679.26 | 9586.42 | 361.28 | 146.61 | 33.20 | ReadLen100.txt | bracken_build | n1000r0 | struo2 | n10r1 |\n",
       "| 42.2097 | 0:00:42 | 7569.01 | 8378.32 | 7557.48 | 7559.79 | 6503.82 | 43.92  | 128.93 | 54.52 | ReadLen150.txt | bracken_build | n1000r0 | struo2 | n10r1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 55.2952 | 0:00:55 | 1536.54 | 1846.98 | 1531.92 | 1532.43 | 2.77 | 0 | 270.36 | 149.79 | RS_GCF_900625105_1_Agathobaculum_sp_Marseille_P7918.txt | vsearch_per_genome | n1000r0 | struo2 | n500r1 |\n",
       "| 40.2421 | 0:00:40 | 1508.84 | 1820.2  | 1504.31 | 1504.45 | 1.8  | 0 |  93.77 |  38.20 | RS_GCF_900638635_1_Bartonella_vinsonii.txt              | vsearch_per_genome | n1000r0 | struo2 | n500r1 |\n",
       "\n"
      ],
      "text/plain": [
       "     s       h.m.s   max_rss max_vms max_uss max_pss io_in   io_out mean_load\n",
       "1    75.4033 0:01:15 7689    8595.88 7677.05 7679.26 9586.42 361.28 146.61   \n",
       "2    42.2097 0:00:42 7569.01 8378.32 7557.48 7559.79 6503.82 43.92  128.93   \n",
       "⋮    ⋮       ⋮       ⋮       ⋮       ⋮       ⋮       ⋮       ⋮      ⋮        \n",
       "7242 55.2952 0:00:55 1536.54 1846.98 1531.92 1532.43 2.77    0      270.36   \n",
       "7243 40.2421 0:00:40 1508.84 1820.2  1504.31 1504.45 1.8     0       93.77   \n",
       "     cpu_time file                                                   \n",
       "1    33.20    ReadLen100.txt                                         \n",
       "2    54.52    ReadLen150.txt                                         \n",
       "⋮    ⋮        ⋮                                                      \n",
       "7242 149.79   RS_GCF_900625105_1_Agathobaculum_sp_Marseille_P7918.txt\n",
       "7243  38.20   RS_GCF_900638635_1_Bartonella_vinsonii.txt             \n",
       "     rule               dataset_start pipeline dataset_add\n",
       "1    bracken_build      n1000r0       struo2   n10r1      \n",
       "2    bracken_build      n1000r0       struo2   n10r1      \n",
       "⋮    ⋮                  ⋮             ⋮        ⋮          \n",
       "7242 vsearch_per_genome n1000r0       struo2   n500r1     \n",
       "7243 vsearch_per_genome n1000r0       struo2   n500r1     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# reading in all benchmark files\n",
    "L = list(\n",
    "    file.path(work_dir, 'n1000r0', 'struo2', 'n10r1', 'benchmarks'),\n",
    "    file.path(work_dir, 'n1000r0', 'struo2', 'n100r1', 'benchmarks'),\n",
    "    file.path(work_dir, 'n1000r0', 'struo2', 'n175r1', 'benchmarks'),\n",
    "    file.path(work_dir, 'n1000r0', 'struo2', 'n250r1', 'benchmarks'),\n",
    "    file.path(work_dir, 'n1000r0', 'struo2', 'n350r1', 'benchmarks'),\n",
    "    file.path(work_dir, 'n1000r0', 'struo2', 'n500r1', 'benchmarks'),\n",
    ") \n",
    "bmk_update = L %>%\n",
    "    lapply(read_bmks) %>%\n",
    "    do.call(rbind, .)\n",
    "bmk_update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-28T11:54:39.121376Z",
     "start_time": "2021-01-28T11:54:14.895Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A tibble: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset_start</th><th scope=col>pipeline</th><th scope=col>dataset_add</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n100r1</td><td>553</td></tr>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n10r1 </td><td>103</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n350r1</td><td>1803</td></tr>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n500r1</td><td>2553</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 4\n",
       "\\begin{tabular}{llll}\n",
       " dataset\\_start & pipeline & dataset\\_add & n\\\\\n",
       " <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t n1000r0 & struo2 & n100r1 & 553\\\\\n",
       "\t n1000r0 & struo2 & n10r1  & 103\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t n1000r0 & struo2 & n350r1 & 1803\\\\\n",
       "\t n1000r0 & struo2 & n500r1 & 2553\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 4\n",
       "\n",
       "| dataset_start &lt;chr&gt; | pipeline &lt;chr&gt; | dataset_add &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| n1000r0 | struo2 | n100r1 | 553 |\n",
       "| n1000r0 | struo2 | n10r1  | 103 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| n1000r0 | struo2 | n350r1 | 1803 |\n",
       "| n1000r0 | struo2 | n500r1 | 2553 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset_start pipeline dataset_add n   \n",
       "1 n1000r0       struo2   n100r1      553 \n",
       "2 n1000r0       struo2   n10r1       103 \n",
       "⋮ ⋮             ⋮        ⋮           ⋮   \n",
       "5 n1000r0       struo2   n350r1      1803\n",
       "6 n1000r0       struo2   n500r1      2553"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# records per dataset\n",
    "bmk_update %>%\n",
    "    group_by(dataset_start, pipeline, dataset_add) %>%\n",
    "    summarize(n = n(), .groups='drop')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-28T11:54:39.190444Z",
     "start_time": "2021-01-28T11:54:16.608Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A tibble: 6 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset_start</th><th scope=col>pipeline</th><th scope=col>n_genomes</th><th scope=col>cpu_hours</th><th scope=col>cpu_hours_wCreate</th><th scope=col>genomes_added</th><th scope=col>total_genomes</th><th scope=col>perc_added</th><th scope=col>cpu_db_create_all</th><th scope=col>cpu_hour_perc_savings</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n10 </td><td> 11.55471</td><td>614.1224</td><td> 10</td><td>1010</td><td> 1.0</td><td> 722.520</td><td>15.00271</td></tr>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n100</td><td> 63.07364</td><td>665.6413</td><td>100</td><td>1100</td><td>10.0</td><td> 782.730</td><td>14.95901</td></tr>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n175</td><td> 83.75373</td><td>686.3214</td><td>175</td><td>1175</td><td>17.5</td><td> 832.905</td><td>17.59907</td></tr>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n250</td><td>119.93707</td><td>722.5048</td><td>250</td><td>1250</td><td>25.0</td><td> 883.080</td><td>18.18354</td></tr>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n350</td><td>180.30615</td><td>782.8739</td><td>350</td><td>1350</td><td>35.0</td><td> 949.980</td><td>17.59049</td></tr>\n",
       "\t<tr><td>n1000r0</td><td>struo2</td><td>n500</td><td>251.80179</td><td>854.3695</td><td>500</td><td>1500</td><td>50.0</td><td>1050.330</td><td>18.65704</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " dataset\\_start & pipeline & n\\_genomes & cpu\\_hours & cpu\\_hours\\_wCreate & genomes\\_added & total\\_genomes & perc\\_added & cpu\\_db\\_create\\_all & cpu\\_hour\\_perc\\_savings\\\\\n",
       " <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t n1000r0 & struo2 & n10  &  11.55471 & 614.1224 &  10 & 1010 &  1.0 &  722.520 & 15.00271\\\\\n",
       "\t n1000r0 & struo2 & n100 &  63.07364 & 665.6413 & 100 & 1100 & 10.0 &  782.730 & 14.95901\\\\\n",
       "\t n1000r0 & struo2 & n175 &  83.75373 & 686.3214 & 175 & 1175 & 17.5 &  832.905 & 17.59907\\\\\n",
       "\t n1000r0 & struo2 & n250 & 119.93707 & 722.5048 & 250 & 1250 & 25.0 &  883.080 & 18.18354\\\\\n",
       "\t n1000r0 & struo2 & n350 & 180.30615 & 782.8739 & 350 & 1350 & 35.0 &  949.980 & 17.59049\\\\\n",
       "\t n1000r0 & struo2 & n500 & 251.80179 & 854.3695 & 500 & 1500 & 50.0 & 1050.330 & 18.65704\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 10\n",
       "\n",
       "| dataset_start &lt;chr&gt; | pipeline &lt;chr&gt; | n_genomes &lt;chr&gt; | cpu_hours &lt;dbl&gt; | cpu_hours_wCreate &lt;dbl&gt; | genomes_added &lt;dbl&gt; | total_genomes &lt;dbl&gt; | perc_added &lt;dbl&gt; | cpu_db_create_all &lt;dbl&gt; | cpu_hour_perc_savings &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| n1000r0 | struo2 | n10  |  11.55471 | 614.1224 |  10 | 1010 |  1.0 |  722.520 | 15.00271 |\n",
       "| n1000r0 | struo2 | n100 |  63.07364 | 665.6413 | 100 | 1100 | 10.0 |  782.730 | 14.95901 |\n",
       "| n1000r0 | struo2 | n175 |  83.75373 | 686.3214 | 175 | 1175 | 17.5 |  832.905 | 17.59907 |\n",
       "| n1000r0 | struo2 | n250 | 119.93707 | 722.5048 | 250 | 1250 | 25.0 |  883.080 | 18.18354 |\n",
       "| n1000r0 | struo2 | n350 | 180.30615 | 782.8739 | 350 | 1350 | 35.0 |  949.980 | 17.59049 |\n",
       "| n1000r0 | struo2 | n500 | 251.80179 | 854.3695 | 500 | 1500 | 50.0 | 1050.330 | 18.65704 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset_start pipeline n_genomes cpu_hours cpu_hours_wCreate genomes_added\n",
       "1 n1000r0       struo2   n10        11.55471 614.1224           10          \n",
       "2 n1000r0       struo2   n100       63.07364 665.6413          100          \n",
       "3 n1000r0       struo2   n175       83.75373 686.3214          175          \n",
       "4 n1000r0       struo2   n250      119.93707 722.5048          250          \n",
       "5 n1000r0       struo2   n350      180.30615 782.8739          350          \n",
       "6 n1000r0       struo2   n500      251.80179 854.3695          500          \n",
       "  total_genomes perc_added cpu_db_create_all cpu_hour_perc_savings\n",
       "1 1010           1.0        722.520          15.00271             \n",
       "2 1100          10.0        782.730          14.95901             \n",
       "3 1175          17.5        832.905          17.59907             \n",
       "4 1250          25.0        883.080          18.18354             \n",
       "5 1350          35.0        949.980          17.59049             \n",
       "6 1500          50.0       1050.330          18.65704             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# cpu hours\n",
    "starting_genomes = 1000\n",
    "bmk_update_s = bmk_update %>%\n",
    "    separate(dataset_add, c('n_genomes', 'dataset_rep'), sep='r', remove=FALSE) %>%\n",
    "    group_by(dataset_start, dataset_add, pipeline, n_genomes, dataset_rep) %>%\n",
    "    summarize(cpu_hours = sum(s) / (60 * 60),\n",
    "              .groups='drop') %>%\n",
    "    group_by(dataset_start, pipeline, n_genomes) %>%\n",
    "    summarize(cpu_hours = mean(cpu_hours),\n",
    "              .groups='drop') %>%\n",
    "    mutate(cpu_hours_wCreate = cpu_hours + 602.5677,\n",
    "           genomes_added = gsub('n([0-9]+)', '\\\\1', n_genomes) %>% as.Num,\n",
    "           total_genomes = genomes_added + starting_genomes,\n",
    "           perc_added = genomes_added / starting_genomes * 100,\n",
    "           cpu_db_create_all = 46.83 + 0.669 * total_genomes,\n",
    "           cpu_hour_perc_savings = (cpu_db_create_all - cpu_hours_wCreate) / cpu_db_create_all * 100)\n",
    "df.dims(20)\n",
    "bmk_update_s %>% arrange(genomes_added)\n",
    "df.dims()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-30T15:14:03.895936Z",
     "start_time": "2021-01-30T15:14:03.510Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyAAAAGQCAIAAADZR5NjAAAACXBIWXMAAB7CAAAewgFu0HU+\nAAAgAElEQVR4nOzdeXwTZf4H8Cf3JGlLzySFQtukgKCCXMq5sMghR5FbxAM8UEG8EFFwXVFR\n0RXwYkVRWFwEueRUORQRFrmqHD85WtqkpQWStPRI22Ryzu+PgWy3LW3SmVzt5/2HL5jOzPP0\nsbSfznznOwKGYQgAAAAA8EcY6gkAAAAANDcIWAAAAAA8Q8ACAAAA4BkCFgAAAADPELAAAAAA\neIaABQAAAMAzBCwAAAAAniFgAQAAAPAMAQsAAACAZwhYAAAAADxDwAIAAADgGQIWAAAAAM8Q\nsAAAAAB4hoAFAAAAwDNxqCcQvsxm84svvujvUW63m/2DSCTie0bBwDCMx+MRCoUCgSDUc2kK\nrH9oYf1DC+sfWlj/0Ar0+mdkZLz++uu+74+AdVMOh+P8+fMZGRnx8fG+H+V0Otk/SCSSwMwr\nsDwej9vtFovFEfoPDOsfWlj/0ML6hxbWP7QCuv45OTlisX+RCQGrEY8++uiwYcN83JlhmGvX\nrhFChEKhX7EsfDgcDovFEhcXF4m/gWH9QwvrH1pY/9BqButvt9srKyux/vWaM2dOWVmZX4eg\nBgsAAACAZwhYAAAAADxDwAIAAADgGQIWAAAAAM8QsAAAAAB4hoAFAAAAwDMELAAAAACe+dcH\nq6CggMtgMTExcXFxXM4AAAAAEP78C1hpaWlcBps7d+4//vEPLmcAAAAACH+4RQgAAADAM/+u\nYH366ad1Nzqdzvfee89oNBJCMjIy+vTpo9PpRCJRXl7eiRMnzp49SwgZNGjQ8uXL1Wo1L5MG\nAAAACGf+Baynn3667sYJEyYYjcaEhISPP/54ypQpQuF/r4oxDLN169bZs2cfOHDg1Vdf3bp1\nK9f5AgAAAPyv8+fPV1dXp6enh3oi/8X1Zc8bNmz47rvvxGLxpk2b/vrXv9b6qEAgGD9+fOvW\nrQcOHLht27a1a9c++OCDHEcEAAAAIIScP38+1FO4Ka41WKtWrSKEjBw5sm668urdu/eYMWMI\nIV999RXH4QAAAKCFO39DqCfSEK4B69SpU4SQu+66q+Hd7rzzTkLIn3/+yXE4AAAAaJkiIld5\ncb1FWF5eTgihabrh3aqqqgghlZWVHIcLJoZhCCFut9vpdPp1CMv3o8KK2+0mhLhcLo/HE+q5\n+A3rH1pY/9DC+ocW1j+gsrOzG97Bu/4MwwRi/T0eT83/xb7gGrBUKlVRUdHRo0cb3u3YsWOE\nkOTkZI7DBRP7FWa1WisqKppwbBOOCh+RFYXrwvqHFtY/tLD+oYX155HBYPD3EIZhArH+LpeL\nDaC+4xqwBgwYsH79+n379m3dunXcuHH17rNt27Y9e/YQQv7yl79wHC6Y2MchFQpFq1atfDyE\nYRiLxcIeGx0dHcDJBYzL5aquro6Ojq75NGikwPqHFtY/tLD+odUM1t/pdFqt1nBYf+/1Krlc\n7uMhDMOwN9MEAkFMTAzvUxKLxSKRyL9DOA75wgsvbN682el0Tp8+3Wg0zpgxQyz+7zldLteX\nX3758ssvs3998sknOQ4XTAKBgBAiEokkEomPh9S8fuj7UWGF/RSa8JUUDrD+oYX1Dy2sf2g1\ng/Vn79uEcP29xVVNmIB3/QUCQSDWXygUsqnAd1wDVq9evd55552XXnrJYrHMmjXrvffeu/PO\nO9u3b08Iyc3NPXbsmPf1hc8880zfvn05DgcAAADNSaQUrfuLa8AihLz44ovl5eVLliyhabqg\noKDuC6GVSuXcuXP//ve/cx8LAAAAmofmGq1YPNxnFQgEixYtunDhwvTp09mX5LDbpVLpLbfc\nMmvWrNzc3IULF4b8ni4AAACEXGR1W2gyHq5gsVJTU1evXk0IcTgceXl5crm8Xbt2CFUAAABA\nmvv1qrp4C1heUqm0U6dOvJ8WAAAAIk5Ly1VefAaswsLCX3/99dSpUyUlJXa7ff369YSQ/Pz8\ntLQ0HkcBAACAMNdic5UXPwHr6tWrL7300jfffFNzIxuw+vTpc9ttt3388ce4rAUAANDsIVqx\neAhYly5d6t2799WrV2+2w08//fSXv/xlz5493bt35z4cAAAAhBvkqlq4FqG7XK6RI0ey6Wrg\nwIE7d+48ffp0zR1eeumlqKiokpKSyZMnu1wujsMBAABA+GghjwQ2AdeAtXnz5rNnzxJCZs+e\n/csvv4wePbpNmzY1d5gzZ86uXbvEYnFeXt6///1vjsMBAABAyCFXNYprwGJbM3To0OGDDz64\nWRf5gQMHZmZmEkK+++47jsMBAABAqCBX+Y5rwMrJySGEjB49WiaTNbAb+5pndmcAAACILMhV\n/uJa5H7lyhVCSIcOHRrerVWrVoSQum/RAQAAgLCFUNVkXANWfHy80WjMzc1teDeDwUBuxCwA\nAAAIZ8hV3HG9RditWzdCyK5duxwORwO7/fTTT4SQzp07cxwOAAAAAsRgMFy4cAHpihdcA9aU\nKVMIIRcuXHj55ZcZhql3nzVr1hw5coQQMn78eI7DAQAAAL/Y+irUSfOLa8B64IEH7rjjDkLI\nhx9++Ne//nXv3r0Wi4X9kMfjycnJeeGFFx555BFCSI8ePWbOnMlxOAAAAOAFHgkMKK41WCKR\naMeOHYMGDdLr9b/++uuvv/4qFF4PbdHR0Varlf2zQqFYs2aNWMz/u6UBAADAd0hUwcH1ChYh\npG3btn/++efbb78dFRVFCPF4POx2b7oaN27cuXPnbr31Vu5jAQAAQNPgelUw8XNJSS6XL1iw\n4LHHHtu6devRo0eNRmN1dXVcXFznzp1HjRo1YMAAXkYBAAAAfyFUhQSf9+zUavVTTz311FNP\n8XhOAAAAaALkqtDieotw6tSpP//8882eHwQAAIBgQul6mOAasNavXz9kyJD09PQ33ngDjdoB\nAABCArkq3PBQ5E4IKSgoWLhwYXp6+rBhw7799luapnk5LQAAADQAuSpsca3BOnHixLp16zZu\n3Hj58mWGYfbt27dv3764uLgHHnjg0UcfZfu888JisWzYsCE7O/vy5ctSqTQ1NbV379733HOP\ntyuEl81m27Rp05EjR4qLi+VyeceOHcePH48m8gAA0GwgUYU/rlewevbsuXTp0kuXLv3yyy9P\nPPFEQkICIaSsrOzTTz/t3r37HXfc8cknn5SWlnIc5dy5czNnzty5c2dOTo7b7bZYLKdOnVqx\nYsWcOXO8fU1ZFotlzpw5mzdvvnz5stvtrqioOH78+Pz589l39QAAAEQ0XK+KFPzcIhQKhYMG\nDfr888+vXr26a9euBx54gO2Jdfr06WeffbZ169ZTpkzZu3evt0WWX1wu16efflpZWXnrrbcu\nX75848aNmzZtmj9/fnx8vF6v//jjj2vuvHz58suXL8fFxS1atGjTpk0rV64cPHgwwzDLly+/\nevUqL58sAABAkOFWYMThJ2B5SSSSUaNGrV271mw2b9iwYezYsTKZzG63b9iwYfjw4WlpaU04\n53/+85+ioiKlUrlgwYK2bdsSQsRicZ8+fRYsWCAQCI4fP37p0iV2z8LCQvalh6+99lqXLl3E\nYrFarX722Wc7d+7sdru3bt3K3ycKAAAQcMhVkYvngOUll8snT568devWwsLCiRMnshsLCwub\ncCqDwUAI6du3b3R0dM3tHTp0SElJIYRcvHiR3fLbb78RQnQ6XUZGhnc3oVA4dOhQQsixY8fQ\nTgIAAMIfclUzEKiXA9rt9r17927evHn79u0VFRVcTnXt2rWYmJj27dvX/ZBcLieEOBwO9q9s\nn4iePXvW2q1Xr16EkLKyspKSkqSkJC6TAQAACBAkquaE54BF0/SePXs2bdq0c+fOmuXncXFx\nEyZMmDp1ahPOOXfu3Hq3m0ymvLw8Qoj3elVZWRkhpG6EiomJkUqlDofjypUr9Qasb775pm4T\nL/b5RJqmq6qq/J0zwzBNOCocsHVyVqtVIBCEei5Nh/UPLax/aGH9Q6sJ65+bmxugyfiFXX+H\nw9HS1t8XLpfL3zpyfgKW1WrdvXv3pk2bdu3aVfMTk8vlmZmZU6dOHTFihFQq5WUslsViWbx4\nsdvt7tq1q/fiFvu4Yq07iayYmJiSkpKbLfqhQ4eysrJqbdRoNIQQp9PZhLZeDMNEdDMwu90e\n6ilwgvUPLax/aGH9Q8v39WcLYMKNy+UK9RQ4CdDXv8fjCXbA2rRp0+bNm7///vvq6ur/nlQs\nHjJkyNSpU8eOHVtv3OGCYZgDBw6sXr26vLw8OTl5zpw53g+x18wUCkXdo9iNkf7vFgAAIl14\n5irgHdeANXnyZO+fBQJBnz59pk6dOnny5ACVOp09e/bLL79k7wz279//ySefbNWqlfejEomE\nEGKz2eoeaLVayU0ubhFCHnjggWHDhtXaaLfblyxZQlEU23LCFwzDsEFTIBAolUofjworLpeL\npmmFQlG3g2v4w/qHFtY/tLD+odXo+ufk5LB/kMlkQZ2Zzzwej9PplEqlEXqLkL2GEqCvf7FY\nLBKJ/DuEl4Fvu+22qVOn3n///U1rxOALq9W6Zs2a3bt3Mwyj0+kefvjhum3ilUpleXl5ZWVl\n3cPZm4Px8fH1nnzAgAF1NxYVFS1ZskQikVAU5eMka/4D8/2osOJwOGialslk/n4lhQOsf2hh\n/UML6x9aN1t/b906ewkgnLlcLqfTKRaLIzTgegNWIL7+hUKhv7mTa8B65ZVXpk6devvtt3M8\nT8PKy8vnz5/PdhB9/PHH+/fvX+/n2bp168uXL1+7dq3WdqvVyt6RZRvNAwAABBSeBwSuKfXd\nd9/1JV3l5eVt2LDh//7v/5owhMPh+Pvf/3758uWePXt+9tlnAwYMuFmK7NChAyHkjz/+qLWd\n3RITE1PzfiIAAAC/DAaDXq9HugISuEajtWzdunXKlCmLFi1qwrG//PJLfn5+RkbGq6++Wm8B\nu1fv3r0JIdnZ2bXeinPgwAFCSJ8+fSL0vjIAAIQztikoqtehJn5qsCwWy4kTJ7Kzs+ttlW6x\nWJYvX04IKS8vb8LJ9+zZQwgZMGBA3Xt/rJiYGPaGa2pqau/evY8ePbpo0aI33ngjMTHR6XRu\n2rTp+PHjYrF4woQJTRgdAACgXrhSBQ3gIWD98ssv48eP9yU8DRo0yN+TMwzDvmpw9erVq1ev\nrnefOXPmeM/89NNPFxQUFBYWPvroo61bty4pKXE4HEKh8JlnnmH7WgEAAHCBXAW+4BqwTCbT\niBEjGu0vJRKJHn744Zo9q3xUVlbmfROOL1q1arVs2bJNmzYdOXKkuLhYLpd369Zt4sSJHTt2\n9HdoAACAmhCtwHdcA9Znn33GpquXXnpp+vTpVqv1/vvvz83N3bx5c8+ePUtLS7///vu33nor\nJibm1VdfbULzj/j4+B07dvh1iEKhmDZt2rRp0/wdCwAAoC7kKmgCrgFr7969hJABAwa8//77\n7JbnnnvumWeeOXPmzIQJE1JTU7t163b77bePHTt24sSJv//+eyR21wAAgBYIuQq44BqwLl++\nTAgZMWKEd0uPHj0IIWfOnPFuuffeewcOHPjrr79+++23TXvfMwAAQBAgVAFfuF5PMhqNhBC1\nWu3dotPpCCHs22y82DfqrF+/nuNwAAAAvDt/Q6gnAs0H1ytY8fHxRqOx5ttpVCqVXC7Pzc11\nu93e1x2wqatuC1AAAIBQQaKCwOF6BSslJYUQcvLkyZobdTqdzWbLzs72bmE7fJaWlnIcDgAA\ngCNcr4Ig4Bqw7rzzTkLIhg0bsrKyvBs7d+5MCPnuu++8W06fPk0IadeuHcfhAAAAmga5CoKJ\na8B67rnnhEIhTdMDBgzwtkofPHgwIWTJkiW//vqrx+PJyspatmwZIaRTp04chwMAAPALchWE\nBNeA1aFDh3fffVcgENA07W1Y9eCDD6pUqvLy8kGDBslksl69erEvB3zxxRe5zhcAAMAHyFUQ\nWjx0pZo3b97OnTvHjx/vfReNUqlcvXq1VColhLhcLkKIQCBYsGDBgAEDuA8HAABwM8hVECb4\nednzqFGjRo0aVXPLyJEjT58+vXLlyvz8/IyMjJEjRw4cOJCXsQAAAGpBooJww0/Aqtctt9yy\nZMmSwJ0fAABaOOQqCFsBDFgAAACBgFwF4Y+3NwMWFBQcPXrU+1en07lgwYK0tLTY2NgRI0b8\n8MMPfA0EAAAtE+qrIILwELCKior69++flpb22muveTfOmTPn3XffLSgoqKio2L1795gxY1au\nXMl9LAAAaGmQqyAScb1FaLfb+/Xrd+nSpZob8/LyPvvsM0KIWq3u2bPnoUOHLBbLc889N27c\nuMTERI4jAgBAs4c4BZGO6xWsFStWsOlq8ODBzz//PLtx27Ztbreboqhjx47t2rXr6NGjSqXS\nZrMtX76c63wBAKD5MhgMFy5cQLqCZoDrFaxt27YRQrp27bp3717vq53ZiqtRo0alpqYSQjp1\n6jRx4sQ1a9YcPHiQ43AAAND8sImK7ZsI0DxwvYJlMBgIIZMnT/amK5fLxVa733333d7dunfv\nTgjJz8/nOBwAADQbKK6CZozrFSyTyUQI8fZwJ4ScOXPGarUSQvr06ePdGBcXRwi5cuUKx+EA\nACDSIVEBj4otEoNJnmekTGWCtx/zhHo6/8U1YLVt2/bixYvFxcXeLQcOHCCExMTE3H777d6N\nZWVlhJCEhASOwwEAQCRCqAK+WKwig1meZ6QMJrneJLdYRd4PlVeXxkeHcGr/g2vA0mq1Fy9e\n/OGHH15++WV2y+bNmwkhQ4cO9d40JDdSV9u2bTkOBwAAEQS5CrizO4UFxZTBRBnMVL6JulIm\nY5j698wpEvfuFC6VfFwD1qRJk/bs2XPw4MH58+dPmzZty5YtR44cIYR4X03IMMyqVat27NhB\nblRiAQBA84ZcBVy4PYJLxTKDWZ53ldKb5FfKZB7fbv1lN6eANW3atHfeeUev1y9evHjx4sXs\nxpSUlClTphBCSkpKunTpcvXqVUKISCSaM2cOx+EAACBsIVdB0zAMMZZL9UZ5npEymOUFxZTT\nJWjCebKLwugFgFynIhaL9+7dm5mZ6f13pVQqP//8c7lcTgix2+3edLV48WKdTsdxuGBiGIYQ\n4nA4bDZbE45twlHhwO12E0JomhYKeXuNUvBh/UML6x9aQV7/ixcv8nUqj8dDCHE6nRG9/oQQ\np9MZ6ik0Bfv1H7T1L6uSGMwKg1luMCkMZrnNIWr8mMbkFIkD9PXvdrs9Pl5Gu4GHrKfT6U6c\nOLF///7jx48nJCSMHj06IyOD/ZBAIOjYsWPPnj1nzJgxcOBA7mMFE7uUdru9urra32MZhmnC\nUeEjQn86emH9QwvrH1rBWX+2R08gRGg68WIYxm63h3oWTRe49bc6xAXFyoJiZX6xsqBEWV4t\n4eW0CpkrLak6NbE6Lal6QI+EAH39hyZgEUKUSmVmZmZmZmat7a1bt75w4QIvQwQfW6QfHR3t\n++t9GIa5du0aIUQoFMbHxwdwcgHjcDgsFktcXFzNZxQiBdY/tLD+oRWc9fferIiKiuL95C6X\ni6ZphUIRiVewvD/XBQKBUqkM9XSagvf1d7gEBWZKb5LrTXK9kTJVSG9WnO4XqdiTprJrNTat\n2qZV0+pYB7mx/vHRcQH6+pdKpWKxf5EpjO5WAgBAGEJlFfjIwwiKrsn0RkpvkucZqaISmYdp\nSilVLUIBk5Jg1yXTWrVNq6FTEuxCAR9JLcAQsAAAoB7IVeALU7lUb6L0JrneKM83yxwufq5+\nqWMdOg2drrZp1bY0lV0qDqMOoj5CwAIAgP9CroKGVVjF3iafeUaqmubnfnqs0pWutuk0dLrK\nptPQSsrNy2lDCAELAKClQ6iCBtgcQsONa1R5Rqq0ip/idLnUcz1RqW06DR0fxam4vrKyMisr\n6//+7//UavXdd99NURQvk+QCAQsAoIVCroJ6Od2CS8UUW0qlN1HGMpmHj5IniYhpl0RrNTRb\nnK6Jswt5KNAiJSUlS5cu/f77773PP0ZFRc2aNevvf/97aJ82QMACAGhBEKqgLo+HXC2T5VyJ\nzr0iLbwWfamEcnv4KE4XktZxdrYyPV1tS02yi4Q8F6fr9frHHnvMZDLV3FhVVfX+++/v27fv\np59+CuETzQhYAADNH3IV1FJskRhM8jwjpTfJ800U7eSnOD0pxqnV2LRqWquxpaloShLA4nSa\npmfNmlUrXXmdPHly6tSpu3fvDtwEGoaABQDQPCFUQU0Wq8hgll+vTzdSFhs/ASBG4WbL0rUa\nW7rKFqMIXnH6hg0bLl261MAOe/bs2bdv39ChQ4M2pZoQsAAAmhW9Xn+z3+mhRaGdwnwzpTdS\nBpM8zyQvruCnOJ2SeNJUtPcyVVJMyDrv//jjj43us2HDBgQsAABoovPnz9fsJB7q6UBouD2C\nS8Uyg1med5XSm+RXymR+vtylfiIh0y7peimVVm1rHWcPk077vryv6dy5c0GYSb14C1gFBQVX\nr17t3bs3+1en0/n666+vW7euvLy8T58+zzzzzMiRI/kaCwAAcAcQGIZcLZOypVQGs7ygmHK6\neOmcTjRxDm9XqlQVLRGFY+d09u3UDfP3BYI84iFgFRUVTZky5fDhw0OGDNm3bx+7cc6cOZ9+\n+in75927d+/bt++zzz6bMWMG9+EAAFoy5KoWrrRSrDfJ2ctUBrPcaufnalJ8lFOrtqXEWzqm\nuLQau0IWAZ3TU1NTG/3n0L59++BMpi6uActut/fr169WlVleXt5nn31GCFGr1T179jx06JDF\nYnnuuefGjRvn+4uTAQCAhVDVklXbRWwdVa6RMpjk5dX83HpSUm5v2/R0tS1W6Yq4l20PGTKk\n0X8a48ePD85k6uL6/2nFihVsuho8ePDzzz/Pbty2bZvb7aYo6tixY2zA7NWrV3V19fLly19/\n/XWuUwYAaAEQqlosh0tQYGabfMr1RspUIWX4uEEnFXvSVHat2sbe+1PHOng4aUg9+OCD69ev\nLykpudkOd91117333hvMKdXENWBt27aNENK1a9e9e/eKRNdfSPTDDz8QQkaNGpWamkoI6dSp\n08SJE9esWXPw4EGOwwEANGMIVS2ThxEUlUj1Jjl7maqoROZheCmlYlIS7Gxluk5DpyTahYJw\nLKVqsujo6E8++eSJJ56orKys+9G0tLRNmzaF8Goc14DF1vBPnjzZm65cLtfRo0cJIXfffbd3\nt+7du69ZsyY/P5/jcAAAzQ9yVQtkKpfqr7/gj8o3Uw4XPzlAE+tg26Zr1bY0lV0qjoBSKi66\ndu26fv36RYsWscGDJRQKp0yZsmzZMpVKFcK5cQ1YbLcVjUbj3XLmzBmr1UoI6dOnj3djXFwc\nIeTKlSschwMAaB4QqlqaCqv4epNPkzzPSFXTIl5OG6t0sbf82DYKSlnw+nyGCa1Wu2rVqvz8\n/GPHjgmFQo1Gk5mZmZycHOp5cQ5Ybdu2vXjxYnFxsXfLgQMHCCExMTG33367d2NZWRkhJCEh\ngeNwAACRC6GqRbE5hAaTXG+i9EZ5npEqreKnz6dC5klT3UhUalt8tIuX00a61NTUxMTE9PR0\noVAYwvcP1sQ1YGm12osXL/7www8vv/wyu2Xz5s2EkKFDh3pvGpIbqatt27YchwMAiCwIVS2H\n0y0ouqa4VBJVVNZKb6KMZTIPHyVPEhHTLonWamid2pautmniHEK0ko0EXAPWpEmT9uzZc/Dg\nwfnz50+bNm3Lli1HjhwhhIwaNYrdgWGYVatW7dixgxDSvXt3jsMBAARUZWUlW9LABUJVC+Fh\nyNVSGdvkM+8qdamEcnv4KE4XkuQ4u05tY6upUpPsImGzKk5vIbgGrGnTpr3zzjt6vX7x4sWL\nFy9mN6akpEyZMoUQUlJS0qVLl6tXrxJCRCLRnDlzOA4HAMA7hmE2bNiwYsWK3377zel0KhSK\nwYMHv/DCC4MHD/b9JAhVLUSJRcJWputNcoOZoh38FKcnxTi9pVRpapqSNPPi9JaAa8ASi8V7\n9+7NzMz0fnNRKpWff/65XC4nhNjtdm+6Wrx4sU6n4zgcAAC/qqurp0yZsmvXLu8Wq9W6a9eu\nXbt2PfPMMx9++GEDj3kjVLUEFqvIYJZfr083UhYbP30+YxRub5NPrYaOkaOUqrnh4QtFp9Od\nOHFi//79x48fT0hIGD16dEZGBvshgUDQsWPHnj17zpgxY+DAgdzHAgDg14MPPlgzXdX0ySef\nREdHv/322zU3IlQ1e7RTmG+m2GtUepO8uIKf4nRK4klTs5XptFZjS4px8nJaCFs8ddxXKjMz\nMzMzM2ttb9269YULF3gZAgCAdzt27GC7Jd/M+++/37dvX61WG7QpQfC5PYJLxTK2jkpvkl8p\nk/HygmCRkGmXZGe7J2jVttZx9gh5Aw3wg1PA2rp16zvvvEMI+fDDD/v168fTlAAAguTLL79s\neAeXy7Vt2zbUjzYzDEOM5VK96fr7kguKKaeLh+J0gYCoW9GpSdUdU1xata1dEi0RoTi95eIU\nsEpKSrKysgghf/zxR9ACltlsnjFjxsMPPzxhwoR6d6ioqNi4ceO5c+cuX74cExPTrl27zMzM\nbt26BWd6ABBBjh8/3ug+Z86cCcJMINBKK8V6k9xbTWW183M1KS7Kyb4yWauxqaOuyaVugUCg\nVCp5OTlENE4Bq1+/fkKh0OPxBPMb0N69e5mbv/fywoULCxcuZFvJS6XS4uJis9mclZU1fPjw\np59+OmiTBIDwd/78+YqKikZ3q/c1ZxD+qu0ivZFi3+5nMMnLq3kqiZG52e4JGRo6XW2LVV4v\nTmcYprq6xXVRhwZw+oLr3Lnz888/v3Tp0q+//vq555677bbb+JpWvZxO5/79+9lGpvWyWq3v\nv/++1Wrt2bPnY4891qZNG7vdvn///i+++GLPnj2dOnXy66FrAGh+apWoJyUlFRYWNnyIWq0O\n5IyANw6XIN9Msa9M1hspY7mUl9NKxZ40lV2rtrGP+2liHbycFpo9ron+gw8+EAPH5hsAACAA\nSURBVIlEH3zwwbhx4xYvXjx+/HiBgP8Ws4cPH96+fXtBQYHNZmtgt3PnzpWUlLRq1eqVV16R\nSqWEEJlMNmLECLPZvGXLlkOHDiFgAbQ0DT/017dv3w0bNjR8BhSYhi0PIyi6JmMf98szUkUl\nMg/DR59PAZOSYGcr03XJdEqCXShAKRX4jWvAWrp0qVqtHj58+O7duydOnJiUlJSampqcnCyT\nyerdf9OmTU0YxWg0mkwmiqIoiqqqqnI663+69dKlS4SQlJQUNl15sW0j2I8CQLPneycF9v0T\nLtdNWxDFx8ePHTuWp3kBD0zlUr2JYlt95psph4ufUip1rIPtnqBV29JUdqkYfT6BK64Ba+7c\nuTX/WlxcXPPFz3yZMGGCt6T9vffeO3z4cL27sVfyL126ZLVaFQqFdzv73RbX+QGaqyb3pkpL\nS3vllVcWLVpU70dFItF7772HguXQqrCKrzf5NMnzjFQ1LWr8GB+0Uri0Ghtbn67T0EoK5VPA\nM64BKy0tjY9p8KN3796dOnU6f/78W2+9NX369LS0tMrKyv379+/cuVMikTzwwAM3O/Cdd96p\n+w06OjqaEGK1WsvLy/2dicfjacJR4YB9gMBisQTiVm/QYP1DKwjrn5eXx9epxo4dKxKJlixZ\nUlVVVXO7SqVauHBht27d2IdmIgjDMBE355rKK50FxYr8YqXBrMg3K8uq+SmlkkvdqUnWtKTq\n1KTqdJU1TlmjlMpDeFywSF9/mqZDPQVOAvT9x+l0ut3+pXCuActgMHA8A49EItEbb7zxz3/+\n88CBAy+99JJ3e2Ji4ssvv9yxY8ebHXjp0qW6AUuj0RBC3G53A7cPGtC0o8KEv19GYQjrH1q8\nr39Av9WMGjWqX79+e/fuPX36tMViSUhI6Nmz59ChQ2UymYeXjpNBF1nTdroFl0sV+WZlfrGy\noERprpB7+Ch5Eos8bROsqYnVbKhStaKFNX5tCegKRdb61xLRk2cF4vs/wzANdDCoFz+PrYaP\nc+fO/fnnn4QQkUikUqkqKyurqqpKS0sPHjyo0+nE4ub2+QI0V0H+5S02Nnby5MmTJ08O5qAt\nFsMIjOUUG6fyi5VF1xRuDy99PpnkWJqNU2lJ1W3irSIhitMhZJpV4Pjjjz/eeustQsjjjz8+\ncuRINk7l5OR8+umnO3fuLCkpmT9/fr0HLlq0yG6319pYXFz8+OOPK5XKuLg4HyfAMAx7ZVIo\nFLZq1arpn0noOJ3OqqqqmJgYkYifQodgwvqHFsf1r/VarZpllEHjdrvtdjtFUQ284DmcsXem\nBAKBXC4P9VxqK7FI2AYKeSZ5QbGcdvCzwkkxznS1VaumtWpbqspGSWpefQn2IoTz+vuieXz9\nB+j7v0Qi8fcaTbMKWKtWrXK73ffff/+YMWO8Gzt06PDaa6/NnDnzyJEjZ8+evfXWW+semJiY\nWHcjezFQKBT6/qOu5vXDSPwBSW7cnBKJRJE4f6x/aPm1/nVvyofD93T25ohQKAyHyfir5vqH\nw/wrbSJvk0+9kbLY+PlxE6Nws2Xp6WqbVm2LUdS6nx6yTzzc1r8Jms3XfyC+fzahLpbrV/yI\nESN833nChAmPP/44xxFvhqZpthFDnz59an0oKSkpIyPj3LlzFy5cqDdgAUCgNflBP4gUdqcw\n30yxT/zlmeTFFRJeTktJPGkqmu1KpdXYkmLqb9MDEG64Bqzdu3f7vnNAW71LJBKpVOpwOOqt\nbmM3huSmA0ALhDjVErg9gsISGds9wWCSXy6V8VIeLRIy7ZLs7GUqrcbWOs4egddTADgHrDvu\nuKPe7S6XKz8/3/vY8z333HPLLbcMGDCA43ANEIlEaWlpOTk5Bw8eZDuLehUWFur1ekJI+/bt\nAzcBgJbs/PnzDMNUV1cTQvCy2+aKYcjVMqmBTVRmeUEx5XTxUpxOkuMc6WpbWlJ1m9iKDimM\nTBrBbUoAWFwD1smTJxv4aGFh4Ycffrh8+fJjx469+uqr/fv35zhcw+67775FixZt375dLBaP\nGTMmNjbW4XCcOnVq5cqVLperZ8+eOp0uoBMAaCFwgarlKK0U601yg1med5UymOVWOz9Xk+Ki\nnDr2XTQaOk1lU8g8hBCXy0XTtESsIAQBCyJeYIvc27Ztu2TJkqFDh44cOXL8+PGnT59OTk4O\n3HC9evV66KGHvvnmm82bN2/evFkul9M0zRa+6XS6559/PqKbNwJwceDAgRUrVhw+fLi8vFyl\nUg0ePNj3F7QjTrUo1XaRnq2jMlIGs7ysip8fE0qZW3sjUaWrbbHK/6nlsFqtP//88++//15a\nWqpWq/v06TNw4MBIfNQDwCsYTxHec889jz766FdffbVs2bL3338/oGNNnDixd+/eW7ZsMRgM\nV65cSUpKSklJ6d+//9133410BS2Tw+GYNWvWV1995d1SVVWl1+tXr1795ptvLliwoNb+iFMt\njcMlKDBTepNcb6L0RrmpQupnP8X6ScVMmorWqm3paptWQ2tiHTfbc9++fW+++ea1a9e8W775\n5pv09PR33323S5cuPEwFIBSC1KZh0KBBX3311Y4dO7gHrJdffrnhHVJSUp577jmOowA0GzNn\nzly1alXd7W63+9VXXy0vL3/kkUeCPysIIQ8jKLomYy9T5RqpohKZh+Hh90+hgElJsHsvU6Uk\n2oWCxpPa9u3bFyxYULdHtsFgmDZt2ldffdW9e3fucwMIviAFLJVKRQgpKioKznAAwNq/f3+9\n6crro48+Gj58eOvWrYM2JQgJU7lUb6L0JrneSOWbKYeLn1IqdayDbfKp1dBpKloq9u8xwsuX\nLy9cuPBmbyCx2+0vvvjijz/+SFEUH5MFCKogBaw//viDEIIHiwCCw3ub77333mt4T4fD8d13\n382ePTvwk4KgqrCK84yU3ig3mOV5Rqqa5qeeKVbpSmfrqFQ2nYZWUpzem7lmzZq6b9GoyWQy\nbd++/b777uMyCkBIBCNg5ebmfvDBBwRdEgACoOGSqYaf82Wxv/9ApLM5hNlXYgpKlPlmZUFJ\nVGkVP30+5VJP+o3KdJ2Gjo/is8/noUOHfNkHAQsiEdeA9dprrzXwUY/Ho9frd+7cyXbHmThx\nIsfhAFqsptWeWyyWRvepqKhowpkh5JxuQWEJlXf1en26sUzm4aM4XSJi2iXd6JyutmniHMKA\nPSBkMpka3cdoNAZqeIBA4hqwFi1a5OOeXbp0mTVrFsfhAFqCCxcu8PUusISEhCtXrjS6Dy9j\nQaB5PORqmYztnpBnpApLKJebj+J0IUmOs7NdqdLVttQku0jIR1LzgUKhoGm60X2CMxkAfgX8\nFqFAINDpdPfcc8/ixYulUmmghwOIIHUvSrGNFnn8iXLXXXdt3bq14X169+7N13DAu2KLhG1J\npTfJ880U7eAneSfFOLUaG1ufnqamKQkf77jx36233troXcKAvmMNIHC4Bqx6X/xXk0AgiMT3\ncgPwKISdpR566KHt27d7bv6KuOjo6HHjxgVzStAwi1XEXqAymOR6I2Wx8fNrcIzCna6yaTXX\n69NjFJyK0/kyfvz4hgOWUCi89957gzYfAB5x/aeLTrsArPDsz3nLLbfMnDlz+fLl9X5UIBAs\nXLgwLi4uyLOCmminMN9MXW+ebpIXV/BTnE5JPGlsAwU1rdXYkmL4LE7ny7Bhw/r163f48OGb\n7TB16tSOHTsGc0oAfAlSmwaA5iE8U1TDZs2aJZVKP/30U6fzf37ERkVFvfHGGyNGjAjVxFos\nt0dQWCLTm65fprpcKrv5FUY/iIRMm3hrWlJ1msraqZ2ndZw9/G8eCASCpUuXzp49+8SJE3U/\nOmbMmEY7SwOELT4DltVqPX/+fG5u7sWLF2ma7tixY8eOHTt37hwVFcXjKACBFokpqgECgWDG\njBn33HPPli1bTp48abFYEhISevfuPWHCBFy7Cg6GIcZyKdvkU2+SFxRTThcPxekCAUmOc3i7\nUrVLtDnsVYQQgUAQQU0Ho6OjV61atXnz5m+//TY7O5sQIhQK77jjjunTpw8ZMiTUswNoOn4C\nlsvl+vzzzxcuXFhSUlLrQ7Gxsa+//vrs2bPFYlwtg3DRzCKUL9q2bfv888+HehYtSGmV2GCS\ney9TWe38XE2Ki3Kyj/tpNXS6yqaQ/ffaF8MwjoZ6doYvkUh033333XfffZWVlUajMSUlRS6X\nh3pSAFzxEHoKCwuHDh3K/uZRV3l5+QsvvPDFF18cOnQIT4ND0LTACAWhVW0XXa+jMlIGs7ys\nip9fKZUyty75etv0dLUtVtnIc0URTS6XJycny2SyUE8EgAdcvwW43e7MzEw2XSUkJDz77LO9\nevVKT0+XSCR5eXlZWVkffvhhcXHx+fPnZ86cuXHjRj7mDEBIfRGKYRi2pW1k3SKBCOVwCQrM\nVJ5Jzj7uZ6qQ3uSVev6Rij1pKjvbkkqnodWxDh5OCgBBxzVgrV279vTp04SQzMzMdevW1Sy3\n0ul0w4YNe/bZZ6dOnbpz585NmzZt2bJlwoQJHEeEFgKXoCDceBhB0TUZe5kq10gVlcg8DB99\nPgVMSoKd7Zyu09ApiXahIEh9PgEgcLgGrE2bNhFC2rVr9+9//7veYvaoqKivv/66a9euly5d\n+vrrrxGwgCA8QeQwV0jZJp96I5Vvphwufkqp1LEOtsmnVmNLU9ml4tD0+QSAwOEasNjXxI4e\nPbpVq1Y32yc2NnbMmDGffvppVlYWx+Eg/CE8QUSz2CTZpuh8s4J94q+K5qfVXyuFS6ehtRob\nW02lpMKizycABA7XgMU+Ntjoqwxuv/12784QuRCeoPmxOYQGk1xvovRGeZ6RKq3ip8+nXOpJ\nv9HkU6eh46PCsc8nAAQO14AVFRVVVlZWWFjY8G5FRUWEkOjoaI7DQYB4kxP7LjyTyYQXHEFz\n5XQLLhVT7AUqvYkylsk8fJQ8SURMuySaLaXSqmlNnF3IQ4EWAEQqrgFLp9NlZWUdOHCg4d1+\n/fVXdmeOw4FfcMEJgBDiYcjVUhnbPUFvlBcUy9wePorThSQ5zq5jW1KpbalJdpEQxekAcB3X\ngDVq1KisrKwjR44sWbLkxRdfrHefjz/++ODBg+zOHIcDgtgE4IMSi8TbOd1gpmgHP1dkk2Kc\nbPcErdqWpqYpCYrTAaB+XAPW7Nmzly1bZrFY5s2bd/Lkyb/97W+33HKL96MXL158++23165d\nSwjR6XQ3S2CAzATAUaVNxDb5ZHOVxcZPn88Yhdvb5FOroWPkzbnPJwDwiOv3oMTExLVr144b\nN87tdn/zzTfffPNNdHS0VqslhOj1+srKSnY3kUj073//G70fa0KoAuCCdgrzzdT15ukmeXEF\nP8XplMSTpqLZu346tS2pFYrTAaApePglLzMz8/jx43PmzGELrSorK9nWo179+vVbtmxZr169\nuI8FAC2W2yO4VCwzmOV5Vym9SX6lTObh4wadSMi0S7Kzb/dLTayKl1colXI85AEAHPFzFb17\n9+4HDhzYtWvXoUOHcnJyLl686PF4OnTo0L59+/79+9977728jBJkHo+HEGK1WisqKvw9lmGY\nRo+y2WxNnFkgMQxDCKFpWiCI7CegwnN5G4X1r4lhiNlCGUyK/GJlfrHiUonc5eYh9wgERN2K\nTlNZ05Kq05KsbROtEhFzY0TG42Gw/qGCr//Qah7r7/F4mvBTu1FOp9Pj5690/AQs1ujRo0eP\nHs3jCcME4/8LxhiGafSoJpw2mMJ8eg3zZf3DXETPn8v6l1VL8s3KghJlvlmZX6ywOfjp8xmr\ndKSrrKmJ1Wmq6rQkq1z6P30+6062xa5/mIjo+Ufu+nunHW7z97EFgct1vUQyQPP397R8Bqxm\nhr1HoFAoYmNjfTyEYZhr166xxzZ6lEKh4DjDQGD7YFEUFYm3SGq+7Dk8l7dRLXD9rXZRHltH\nZaQMZnlZFT/flJQyN1tHlaGh09W2WGXN4nTZzY5qgesfVrD+oRWc9e/UqVMgTuvXz98mkEgk\nIpF/v+8hYAFAUDlcggLz9SafepPcVC7l5bdNqdiTprJr1Tb2cT91K0ck3+UAiBgBCkzNAA8B\ni2GYzz//fO/evdnZ2Y3eocSjcwAtjYcRFJVI2SafeUaqsETmYfjo8ylgUhLsbOd0nYZOSbQL\nBeF1XwMgEt1yyy3+XqqBenENWB6PZ/To0T/++CMvswGA5sFULmUvUOmNVL6Zcrj4ueOgjnVc\nb0mltqWp7FIx+nwCNM7Hi0x2u93bXAm44xqw1q5dy6YrgUDQoUOHDh06IPkCtEAVVnGekcou\njC0oVhqKlVY7P+UHrRQu9mXJbLdPJeVu/BiAlgH35sIc12+Cq1atIoRQFLVly5aRI0fyMSUA\niAA2h9DAXqMyyfOuUqVV/PT5lEs97Lto2P/GR6HPJ7QsiE3NBteAlZOTQwiZOXMm0hVA8+Z0\nCwpLqDwjpTfK9SbKWCbz8FHyJBEx7ZJotpRKq6Y1cXYhitOhOUJyamm4BqzS0lJCyKBBg3iY\nCwCEEw9DrpbK2O4JeUaqsIRyuXkpTifJ8Xb2rp9WY0tNsouEKE6HCMYmp5ptAuLj40M9KQg9\nrgFLo9EUFBTQNM3LbAAgtEosErYyPc8kzzdTtIOf4vTEGCf7Lhqd2pamoikpitMhMuCyEzQZ\n14A1fPjwL7744siRI5MnT+ZlQgAQTJU2kZ5t8mmS642UxcZPcXqMws2WpbNdqWLkrsaPAQg6\n5CcIHK7fTOfNm7du3boVK1Y89NBD3bt352VOABA4tFOYb6b0bPN0k7y4gp/idEriSVPRKfGW\ntKTqdLW1nRpNjCH0kJ8ghLh+E9TpdN9+++2kSZNGjBjx5ptvPvbYY2IxvrEChBG3R1BYIvNe\nprpcKvPzjaX1EwmZdkn2dJVNl0xr1bbWcXaB4L+vCsFbIiAIkJ8gnPn3TXDSpEn1bm/Xrl12\ndvZTTz01d+7c9u3bp6am3ixmbdq0ye85AoA/GIYYy6VsKZXBJM8vppwuHorTBQKiiXVo1TZd\nMp2usqUm0RLx/xSnh9n7YaE5QISCyOVfwNq8eXPDO1RVVZ08efLkyZMcpgQAfiutFOtNcoNZ\nnneVMpjlVjs/xelxUc7rTT6T6XSVTSFDcTrwrGaEcjgcFoslLi4ODauhGfAvYKWlpQVmGgDg\nn2q76HodlZHSm+Tl1fzcklPK3GyWYuvTY5UoTgce4EIUtED+fVM2GAwBmgcANMzhEhSYr3dO\n1xspU4WUl1tyUjGTpqK1ahv7uJ8m1sHDSaFFQooCqAmFqABhysMIikqkBrNcb5TnGqmiEpmH\n4aXPJ5OSYGc7p+s0dEqiXShA8RT4AUEKwBcIWABhxFwhzSlSFJXHGkzygmLK7uSnlEod69Cq\nabbVZ5qKlopRSgWNQIoC4AgBCyCULFYxW0SlN1J6k7yK5qe2t5XCpdPQWo2NraZSUm5eTgvN\nD4IUQIAgYAEElc0hNJj+W0p1rZKfPp8KmSftRmW6TkPHRzl5OS00J8hSAMGEgAUQWE634FIx\nxV6g0psoY5nMw0fJk0TEtEui2bf7pattmjiHkIcCLWgOtFotXjYMEHIIWAA88zDkaqksz0gZ\nzPI8I1VYQrncfBSnC0lynF2nuf7EX2qSXSREcXpLV/OiFMMw165dC+FkAKAmBCwAHpRYJGyT\nT71JbjBTtIOf4vTEGCf7rJ9WbUtT0ZQUxektF27wAUQWBCyApqi0idgmn2yustj4+acURblS\nk6o7tHZoNbRWQ8fI0eezhUKcAoh0CFgAPqGdwnwzdb15ukleXMFPcTol8aSpaC1bnK62xSlt\nNE0rFAqhkJ9rYBD+kKUAmqXIC1hms3nGjBkPP/zwhAkTbrbP0aNHf/zxR71e73A42rZtO2DA\ngNGjR+PlVuAXt0dQWCLTs5epTPLLpTIPHzfoREKmXZKd7Z6g1dhax9lrRikXrlg1d4hTAC1E\n5AWsvXv3Mg2+IuSf//zn7t27CSECgUAkEuXk5OTk5Bw5cmTRokViceR9vhA0DEOM5VK96fr7\nkguKKaeLh+J0gYAkxznY7gnpKltqEi0Rozi9pUCcAmixIilwOJ3O/fv3b968uYF9du3atXv3\nbolE8tRTT/Xu3VupVJ4+fXrZsmXnzp1bt27dww8/HLTZQkQorRIbTHLvZSqrnZ8bc3FRTjZO\nsS9OVshQnN4iIE4BgJd/ASszM9P3nePi4nQ63dixY7t27ernrGo7fPjw9u3bCwoKbDZbA7vR\nNL1+/XpCyNy5c/v06cNu7Nat2+zZsxctWvTTTz899NBDAgGaBbVoVrtIb6L0xuv16WVV/PyO\noZS5tTeafKarbHFRuNXX/CFOAUAD/PvpsmvXLn8HWLhw4aRJk1asWMGl8Z3RaDSZTBRFURRV\nVVXldNbfpfrEiROVlZWtW7f2pitWr1693nrrLUKI0+mUSqVNngZEIodLUGC+3uRTb5KbyqUN\n3mH2lVTMpKpordrGvuBP3cqB6N7sIVEBgO+CcYtw06ZN2dnZ//nPf6Kjo5t2hgkTJnhL2t97\n773Dhw/Xu9vZs2cJIf369au1XSAQcL+KBpHCwwiulimuGmINJnmukSoqkXkYPvp8CpiUBLtW\nQ7ONqVIS7UIBSqmaM8QpAODCv4CVm5vr454MwxgMhmPHjn3wwQcVFRVnzpz54IMP3njjDf9n\n6IeSkhJCiEqlMpvNu3fvzs7OLi0tTU1N7dat27Bhwxq4OWi1Wl11Ht+yWq3sJ9JwTX1NNfds\n9CjfTxsSYT69WswVUraOKu+q7FKJwuHip5RKHevwPu6XmkRLxf9TShXQFYqs9a8rEuffqVMn\nh8NRWVkZGxsrEoki7lPw6/tPeGKn7dd33fCB9Q+tMFx//wKWTqfzfeeMjIyhQ4dOmTJlxIgR\nubm5S5cuffrpp1UqlZ8z9AP7moji4uK5c+eWl5ezGy9fvvzbb7/9/PPPf/vb32JiYuo9cM6c\nOVlZWbU2ajQaQkhVVVUT3j7h8XgaPaq6utrf0wYNGy7DmcUmyS9WFhQrC4qVhmKl1c7PtdgY\nuTMtqTpNVZ2aWJ2WVK2Q/Td2O+3EaedlkMaF//o3jGGYcP7y9kpPT6/5V++/We93jwjly/ef\ncIb1Dy2sf70cDkfdCzENC/gtwoyMjH/84x/jxo2rqqpavnx5QC9isWu6cePGtm3bPvPMMx07\ndnS73SdPnly9evWFCxdWrlz54osvBm50CCjaKbpUcj1OFRQrS6v4qaWjJO7UpOrUxOo0VXVa\nUnWc0sHLaSE81UpUAACBE4warLFjx2ZkZOTm5h48eDCgA7HNr+Pi4hYvXuyt9xo8eLBKpVqw\nYMGvv/46adKkdu3a1T0wMTGxTZs2tTbGxsYajUahUOhXh1K3283+odGjwrNVN3txOBzm5nQL\niq4pCoqVBrOioERpKqc8fFz0lYiYlARralJ1usqamlitjqWF/3PrOMSfePisf9N4bjRjDZNP\nQavV+rU/wzAejydymxL7/v0nPGH9Qwvr3wCBQOBvF4Ig9cG67bbbcnNzs7OzAzpKfHx8aWnp\n4MGDa1XT33bbbSkpKUVFRXl5efUGrEWLFtXdWFRUNHbsWKVSGRcX5+MEvG+zFwqFjR6lUCh8\nPG0wuVwumqYpigr+D0iPh1wtk7HdE/RGeUGxzO3hozhdSJLj7Gz3BK3GlppkFwm9SU1ISHj9\nXwjh+nPnvTMoEAhC9eXNsTLd4XBYLJaYmJhI/Bnj1/ef8IT1Dy273V5ZWYn1r5dEIvF3WYIU\nsNjfI0tLSwM6CtsJQq1W1/2QSqUqKioqLi4O6ATAXyUWid4k1xupPJM830zRDn5SRUKUXZds\n12lordqWpqYpCfp8Nlt41g8AwlOQApbdbieESCT8vB/3ZlJSUo4fP37lypW6HzKZTISQtm3b\nBnQC0KhKm8hgkucZKTZXWWw8Facr3Okqm1Ztax1bnppUHSN3KZVKXs4M4QaJCgAiQpACVn5+\nPrlxhSlwhg0btnXr1l9++WXSpEk17xKePn368uXLQqGwffv2AZ0A1GV3CvPNFPsimjyTvLiC\nn5BNSTxpKvp683S1LamVk/zPw2to+tmsIFQBQMQJRsCiafr48eOEkC5dugR0oNatW3fr1u2P\nP/549dVXZ82a1aFDB5fLdfz48c8//5wQMnbs2MTExIBOAAghbo+gsETmfbvf5VKZh48bdCIh\n0zbRzjb51GpsrePsEVikBL5CogKASBeMgPXll1+yxU9Dhw4N9FizZs165ZVX8vPz582bJ5VK\nXS4X+1hTr169pk6dGujRWyaGIcZyKXvLz2CS5xdTThcPF5AEApIc5/C+3S81iZaIw6J3HAQI\nQhUANCcBD1gbNmx4+eWXCSFt2rSZMWNGoIdTqVQff/zx5s2bjx07VlxcHB8fn56ePmjQoAED\nBgR66BaltEpsMMlvXKairHZ+HjmJi3LqbryLJk1lU8hQnN6cIVEBQDPmX8D66KOPfNyTYZj8\n/Pzjx48fOXKE3bJw4UK5XO7f7OrDxrUGREVFTZ8+ffr06dzHAi+rXXS9jspIGczysip+orlS\n5mbrqDI0dLraFqv0r08uRByEKgBoIfz7Mfn88883bZj777//kUceadqxEBIOl6DATOlNcr2J\n0pvkpnIpLy93koqZVBWtU9vS1TathtbEonN684dQBQAtUMBvEWo0mtdee+2pp56KxMaJLYrH\nQ4quyXKvRudekV26Fl1YIvMwfPT5FDApCXbtjRt/KYl2oQClVM0fQhUAtHD+BawVK1b4vnNM\nTEx6evqdd96JaBW2zBVSvfH6Zap8M2V38vN/StXKwZZSaTV0moqWilFK1SJ06tTJ20k51HMB\nAAgx/wLWk08+GaB5QHBUWMVsKRWbq6poforTWylcOg3N3vXTqm1RlJuX00L4q3mliuHlLjIA\nQLMQpEajECo2h9BgYtumy/OMVGkVP30+5VJPutqmVdNajU2noeOjnLycG8xXtQAAIABJREFU\nFsIf7v0BAPgCAau5cboFhSVUnpHSG+V6E2Usk3n4uKwgETHtkmj2ApVWTWvi7EI0S29JkKsA\nAPzSxIB18ODBPXv2/Pnnn1euXLn11lu7dev2yCOPxMTE8Ds58IWHIVfLZHlXKYNZnmekCkso\nl5uX4nSSHG9nm3xqNbbUJLtIiBtALQtCFQBAk/kdsCorK+fOnfvFF194t2RlZa1Zs2bJkiUr\nV64cPnw4r9OD+pVYJAazPO8qpTfJDWaKdvBTU5wY40xXWVPiLbe0dWvVdkqK4vQWB6EKAIAX\nfgesmTNnfvPNN3W3FxYWZmZmZmVlBfqFgy1TpU3kfbuf3khZbPzc242Wu9nuCWx9eozc5XK5\naJpWKBR4EKzlQKgCAOCdfz+njxw5wqYrgUDwzDPPjBs3TqVSnT59+o033sjOznY6nY8++ujx\n48fxs5k72inMN1Ps2/3yTPLiCn6K0ymJJ03lLaWyJbVCcXoLhVAFABBQ/gWszz//nP3D8uXL\nZ86cyf65c+fOY8aM6dGjR3Z29u+//37hwoXOnTvzPM0WwO0RFJbIvJepLpfKPHzcoBMJmXZJ\n9nSVTZdMa9W21nF2pN8WC6EKACBo/AtYOTk5hJCMjIxaDbGUSuXrr78+depUQsjp06cRsHzB\nMMRYLtXfaElVUEw5XTwUpwsERBPrYLsnpKtsqUm0RIzi9BYNuQoAIPiaErDqbc5+1113sX/I\nzs7mZWbN2x8XyfMrOljt/PT5jItysnFKl0ynq2wKGYrTWzqEKgCA0PIvYLHvwWjXrl3dD6Wm\nprJ/cLvRxbtxyfGES7pSytxaDZ2uvn6ZKi7KxePcIEIhVAEAhI+mPIwmFtdzlEjEz8WYFiI5\ngcQoXBarr+svFXvSVHat2sY+7qdu5RCgzycQQpCrAADCEjq5h4xOQ5/UR93so0IhSY61p6np\ndBWdrqbT1TaJCKVUcB1CFQBAmEPACpl0la1WwFLHOrRqWqu2aTW2NJVdKkYpFfwP5CoAgEiB\ngBUyOg2tkLnT1XTH1rZUlS1DY4tRoHwNakOoAgCIRAhYwVZUVLRhw4ZTp04Zjcb01q0H3jGw\n9129Baiogv+FXAUAENGaErB27dplNBqb9tGVK1c2YcTmgWGYt956691336Vp2rvx66+/vuOO\nO95///2UlJQQzq3ZMJlMFRUViYmJSqUy1HPxW6dOnRwOh8ViiYuLC/VcAACAk6YErFOnTp06\ndappH23JAevpp5/+7LPP6m4/derU1KlTv/3229atWwd/Vs0DTdNff/31hg0brl69ym7R6XTT\np08fN25c+L+4CRerAACaH9wiDJLt27fXm65YJSUlr7766urVq4M5pWbj2rVrTz755Llz52pu\nzMvLe+211/bv37906VKZTBaqud0MQhUAQPPmX8A6evRogObR7C1evLjhHY4dO3bmzJkuXboE\nZz7NhtvtfvbZZ2ulK69ffvnlzTfffPvtt4M8q5tBrgIAaCH8C1je9+GAX8rLy48dO9bobocO\nHULA8teuXbtOnjzZwA7btm27//77b7vttqBNqS7kKgCAlga3CG+KfedPVVUV+4Igv3g8nppH\nZWdnM0zjbUIvX75cXV3t71iBYLPZQj0FX3333XcN78AwzJYtW9LT04MzH6+aI/r7JVReXs73\ndIKq1td/xMH6hxbWP1TYn1NY/3o5HA5/3wSIgHVTbHG0RCKhKMr3o9hoIhAIah6VkJDgy7HR\n0dH1voYomBiGcblcIZ+G79gXkDe6T9A+o/bt23M53OPx2O12mUwWoZ076v36jyBY/9DC+oeW\n2+12OBxY/3qJRCJ/l4XrT53Kykq5XF7rp5der1+7du2QIUPuvPPOCPpRXQu7lDKZzPcH/hmG\n8f4PrnlU+/btk5OTvQ+43UzXrl1DXo7tcrlcLpdEIgn/h+9Ydru90X3YbxkBnQZfNwEdDofd\nbpfL5ZH4cs+bff1HEKx/aGH9Q8tutzscDqx/vUQikb8/Fpv4Q9Tj8WzcuHHixIkqlarutbiy\nsrLXX3+9X79+CQkJM2fODJPbXiEkFAofeeSRhveJj48fNGhQUKbTrPjS26JNmzaBGLpTDYE4\nPwAARK6mBKzS0tJ77rnnvvvu27JlS82emXVZLJYVK1Z07979999/b+oMm4l58+Y1fPPolVde\nUSgUQZtPs+FLKh04cCCPIyJUAQBAo/wOWAUFBT169Ni3bx/713bt2tW9+dK+ffsPPvhg8ODB\nUqmUEJKTk9O3b9+srCzu041crVq1+vHHHzt27Fj3QyKRaP78+aNHjw7+rJqBadOmNRxM27Zt\ny8vaIlcBAIDv/A5Y8+bNy8/PJ4R07tz5559/zs/Pj42NrbVPTEzMiy+++PPPP589e7Zv376E\nEIfD8dBDD3k8Hj7mHKl0Ot3vv//+9ttvey9lyeXy4cOHb9y48aGHHgrt3CJXUlLS4sWLb1Yx\nEBUV9eGHH0okkiafH7kKAACawL+Adfz48Y0bNxJC+vTp89tvvw0ePLjhovqMjIyDBw8OGzaM\nEHLhwoWdO3dymWszoFQqFyxYkJOTU11dfejQoaysrGXLluGHN0dDhgxZuXJl27Zta22/7bbb\n1q1b14TlRXEVAABw5N8jfv/617/YPyxbtqxVq1a+HCISiT766KMuXbo4nc4dO3bce++9/k6x\nWVIoFD72bgBf9O7de9euXYcPHz5+/LjFYklMTOzfv3+PHj38eqoWcQoAAPjiX8C6ePEiIaR7\n9+5+tXS/5ZZbevTocfToUbxpBwJHIpEMHDiwZ8+exM/HdJGrAACAd00JWB06dPB3mPbt2x89\netRoNPp7IECAIFcBAEDg+Bew2ITUhK5C7CGVlZX+HgjAL+QqAAAIAv8CVlJSUlFRURMah5aW\nlhJCEhMT/T0QgBfIVQAAEEz+Baw2bdoUFRX9+eef/g5z6tQpQohGo/H3QIAmQ6gCAIBQ8a9N\nw5AhQwghv/32W0FBge9HGQyGEydOEEL+8pe/+DUcQBOkp6drtVqkKwAACCH/AtakSZMIIR6P\n54UXXvD9qJdeeolhGEIIejRA4LBtq9LT00M9EQAAAD8DVteuXSdMmEAI2bp16/PPP+92uxve\n3+Px/O1vf9uyZQshpE+fPn/961+bPFGAeqEjKAAAhCG/X5Xz4YcfsrXqH330UefOndetW3ez\nF+D8/PPPPXr0ePvttwkhMpnsk08+4ThXAC/kKgAACGf+FbkTQlJSUn766afBgweXlpbm5OQ8\n8MADf//733v06KHT6bRabXx8fHl5+cWLFw8cOOBtKyoUCtetW9ejRw++Jw8tDhIVAABEBL8D\nFiGka9euWVlZjz322C+//EIIycvLy8vLu9nO7dq1+/rrrwcOHNj0OUKLh1wFAACRxe9bhKz0\n9PSff/55x44dw4cPF4lE9e6j0+mWLl36559/Il1B0+A+IAAARKimXMFiCQSCzMzMzMzMqqqq\no0ePXrhwobS0lKbp+Pj45OTkvn374nkuaBokKgAAiHRND1heUVFRQ4YMYVtkATQZchUAADQb\nPAQsAC6QqwAAoPlBwILQQK4CAIBmDAELggq5CgAAWgIELAgG5CoAAGhRmtimIYTMZvO9997L\nvn6nUS6Xa86cOWPGjDEajYGeGNSFPgsAANAyRd4VrL1797KvjvbFv/71r9zc3IDOB+pCogIA\ngBYukgKW0+ncv3//5s2bfdz/+PHjO3bsCOiUoCbkKgAAAFZkBKzDhw9v3769oKDAZrP5eEhJ\nSclHH32kUCisVmtA5wbIVQAAALVERg2W0Wg0mUwURcXFxUkkkkb393g8S5YsqaysnD17dhCm\n1zKhvgoAAOBmAnsFi6bpa9eutWnThuN5JkyYMGHCBPbP77333uHDhxvef/369WfPnh0xYkTf\nvn05Dg21IFEBAAA0KlABKzc394UXXtizZ4/T6UxMTFy4cOHTTz8doLFqOXPmzMaNG1NTUx97\n7DEfD/njjz9KS0trbayuriaEOJ1Ou93u43m81fcMwzR6lMvl8vG0weR2u9n/ejyemts7dOjA\n/sH31Qg+v9Y/PLFfFQ6HQyiMjKvLNWH9QwvrH1rNYP2dTifB+t+Ex+Px/QE7VkAC1o4dO+67\n7z6aphUKhUwmKykpmT179qlTp1auXBmI4WqqqKhYsmSJVCqdN2+eVCqtlRJu5osvvsjKyqq1\nUaPREEJomq6srPR3GgzDNHoUTdP+njZovF+d3jd2N2ERQsiX9Q9nbLiPXFj/0ML6hxbWP7QC\ntP4ul4u9AOE7/lPq2bNnH3jgAbfb/cUXX1RUVJSXl69evVoqlX755Zfff/8978PVxDDMsmXL\nysrKnnjiibZt2wZ0rOYt/YZQTwQAACAi8X8Fa/78+VVVVfPmzZsxYwa7Zfr06Tk5Oe++++6K\nFStGjRrF+4he33333R9//NG/f/+hQ4f6dWD37t1jY2NrbZRIJD/++KNYLJbJZL6fir32IxAI\npFJpw3uKxWH3CGdGRobH43E6nVKpVCAQhHo6TeH7+ocnrH9oYf1DC+sfWlj/BgiFQn/vnPL/\nM/7gwYOEkHvvvbfmxrFjx7777rsnTpzgfTiv/Pz8tWvXqtXqJjw5+MQTT9TdWFRU9OOPP8rl\n8ujoaB/P4731KxAIGj2Koih/5xkgNevWHQ6H0+lUKpWi/2/vzuOiKvc/gD9nZphh0WGXRWQR\nQQQzMXfBFW+4oARYll3TtFAzzQVI3Ius9CpiomaWpeaOiLlQXqByJ6+ahLxAKRABWUb2bWBm\nfn88957fvGYGkuEMw+jn/Rc8y3me+R5gvjznnGf4fD1OSTvtin/XhPjrF+KvX4i/fjU1NSH+\nrREIBDpPsE6cODFjxow2GtAZqNwLRm9dfJodFrRWVFQkk8lKSkpmzpypXktTqCFDhqxdu1Z3\nczAseB4QAABAR9p9D9arr746evTo27dvt9YgICCAEHLs2DHlQvqtr69v+2f4tAQCgZkaU1NT\nWmtiYmJmZtZ1Fo30CPtXAQAA6Fq7V7BCQkJOnTo1ePDguXPnfvLJJ3Z2dioNPv/887S0tPj4\n+J49ey5atEihUOzZsyc+Pp4QsmjRIm5mrcnQoUOPHDmiUiiXy4ODgwkhcXFx9KnA5xmSKgAA\ngM7R7hWshISE1NTU/v37f/31156enlu2bJFKpcoN3NzcEhISxGLxhx9+aGFhYWlpGRUVJZPJ\n5s6dGxgYyN3M4WlhyQoAAKCTabNNw7hx427durVnzx6RSBQZGenj45OUlKTcYPTo0dnZ2XPm\nzDE3N5fL5dbW1jExMV9//TVHc4angrwKAABAX7TcB4vP54eHh+fk5Cxbtiw/Pz84OHjixIl/\n/PEH26BHjx779++XSCQFBQVlZWWrV6820Mc+DQ7yKgAAAL3r0DYNFhYW27ZtW7BgwfLly8+d\nOzdw4MDw8PCPPvrI2tqaNmAYxsnJiYt5/r+oqKinb8zj8c6cOcPtBLomZFQAAABdBwc7uXt6\nep49e/bChQuenp67du3y8PDYsWNH1/ygvWcP1qsAAAC6IM4+KicwMPDu3btxcXGEkKVLlw4Y\nMCA5OZmrg4MK5FUAAABdWYcSrKqqqsuXLz958oR+KxAIlixZcv/+/ffeey8nJ2fSpElTpkzJ\nzs7mYp5ACPIqAAAAA6FlglVdXf36669bWVn5+/vb2Ni89tpr1dXVtMra2nrnzp137twJCAg4\nf/78Cy+8sHz58srKSu7m/DxCXgUAAGBAtEmwpFKpn5/f0aNHFQqFi4sLwzDHjx/38/NT3hCr\nf//+Fy9eTEpKcnFxiY2N9fDw2LNnj0wm427mzwUsWQEAABgibRKsTz/9NCMjw8XF5d69e3l5\neVlZWa6urhkZGZ9++qlKy2nTpmVmZm7evFkqlS5cuHDQoEFczPnZh7wKAADAoGmTYB06dIgQ\nsnXrVi8vL0KIp6fnv/71L0LIwYMH1RsLhcKIiIj79+/Pnz9feaMsUIe8CgAA4NnQ7n2wamtr\nc3NzCSHjx49nC+nXubm5lZWVFhYW6r169Ojx1Vdf6fSzCA0XMioAAIBnTLsTLIFAwOfzW1pa\nlO+4amxsJITw+XyRSNRGX19fXy2m+KxCXgUAAPCsavclQmNjYx8fH0LI6dOn2cLExERCiLe3\nt4mJCYeTAwAAADBE2tyDtWzZMkJIVFTUqVOnamtrExMTV61aRQhZvnw5x7MDAAAAMEDafBbh\nW2+9debMmVOnToWGhrKFwcHBc+bM4WxeAAAAAAZLy41GT548uWnTJmdnZ0KIk5PTxx9/nJCQ\nwOnEAAAAAAyVNitYhBCGYVatWrVq1arWHhsEAAAAeG519MOekV0BAAAAqOhoggUAAAAAKpBg\nAQAAAHAMCRYAAAAAx5BgAQAAAHAMCRYAAAAAx7TcpuH5ceTIkdTU1Kdv39TURAhhGEYoFOps\nUjokl8ubm5uFQiHDMPqeizYQf/1C/PUL8dcvxF+/dBr/rKwse3v7dnVBgtUqoVDYr1+/lpaW\nwsLCp+9VV1dHCGEYxtTUVGdT0yGFQiGXy3k8noH+giH++oX46xfir1+Iv37pNP42Njaurq7t\n6sIoFArO5/Hcqq2tHTt2LCHExsYmOTlZ39N57iD++oX46xfir1+Iv37V1NSMGzeOEGJra3vh\nwgV9T4cQ3IMFAAAAwDkkWAAAAAAcQ4IFAAAAwDEkWAAAAAAcQ4IFAAAAwDEkWAAAAAAcwz5Y\nXBKJRNHR0YQQY2Njfc/leYT46xfir1+Iv34h/vplbGzc1eKPfbAAAAAAOIZLhAAAAAAcQ4IF\nAAAAwDEkWAAAAAAcQ4IFAAAAwDEkWAAAAAAcQ4IFAAAAwDHsg8WNhoaGEydOXLt2rayszMTE\npG/fviEhId7e3vqe17OstLT0nXfemT17dmhoqMYGOCm6UF1dfezYsezs7MLCQqFQ6OLiMnz4\n8MDAQB5P9b81xF8X/vzzzxMnTjx8+LC0tNTKyqpnz57Tpk0bOHCgekvEX9daWloiIyMfPHiw\nd+9ee3t7lVrEn3MXL1784osvNFZNmDBh6dKlyiVdIf78DRs2dOZ4z6Tq6uqIiIgbN27U1NQQ\nQhoaGgoLC1NSUmxtbXv37q3v2T2zEhMTMzMzBw4cqPF3BidFF+7duxcVFZWRkSGRSPh8fkND\nQ3Fx8c2bN9PT00eOHCkSidiWiL8unD179tNPP3348GF1dbWJiUlFRUVRUVFaWlpVVdXgwYOV\nWyL+neCbb765du0aISQoKKhbt27KVYi/Lly6dCkrK0tjVe/evYcPH85+20XijxUsDsTHxxcW\nFlpaWq5YscLb21sikRw5ciQ1NTU+Pt7Hx8fBwUHfE3zWNDc3p6amnjx5so02OCmca2lp2blz\nZ01NjY+Pz6JFi3r16tXS0vLbb799+eWXf/75544dO9asWcM2Rvw59/jx42+//VYul4eFhYWF\nhZmamjY1Nf3www8HDhw4f/78oEGDhg4dyjZG/HUtPT39zJkzrdUi/rpQWFhICJk/f77yjzpl\nYmKi/G0XiT/uweqogoIC+k/M2rVrBwwYIBAI7OzslixZ4u3tLZPJEhMT9T3BZ8qVK1ciIyPf\nfPPN+Ph4uVzeWjOcFF24fPnyo0ePzMzMoqOje/XqRQgRCAQjRoyIjo5mGCY9Pf3hw4e0JeKv\nC8nJyVKpdNCgQbNnzzY1NSWEiESisLCw0aNHE0KuXr3KtkT8da28vDwuLo6eBXWIv44UFRUR\nQl544QV7Nebm5myzrhN/JFgdRf+uubu79+nThy3k8XgTJ04khNy4cQMfRsShx48fl5SUGBsb\nW1paGhkZtdYMJ0UX/vrrL0LIyJEju3fvrlzu6enp5ORECLl//z4tQfx1geavQ4YMUSn38fFh\naynEX6fkcvnWrVtramoWL16ssQHirwtyuby4uJjH49G/Nm3oOvHHJcKOys/PJ4So3ABB/vd3\nsKKiory83NbWVg8zexaFhoayt7R//vnnV65c0dgMJ0UXJBKJWCz28PBQr6Lr81KplH6L+OvC\n2LFjX3rppZdeekmlvKqqihAiFovZEsRfp44cOZKZmTlp0qSRI0dqbID460JpaWlLS4uTk1N2\ndvbZs2cLCgqEQqGrq+vIkSNV/uvoOvFHgtVRFRUVhBD1syUWi4VCoVQqLSoqwu9SJ8NJ0YWV\nK1dqLC8pKcnNzSWEsP8vIv66QC8FKpPL5bdv305KSiKEvPzyy2w54q87d+/ePX78uIuLy7x5\n81prg/jrAr0Bq7i4ODo6mi3Mzc1NSUnx9/f/4IMP2GsaXSf+SLA66smTJ4QQlYsmlFgsLi8v\nr62t7fRJPe9wUjpNdXX1Z599JpPJXnzxRXZxC/HXtSNHjqSlpVVWVjY2Npqbmy9ZsmTEiBFs\nLeKvI1VVVVu3bhUKhZGRkUKhsLXbQBF/XaAJlkwm+8c//hEQEODi4lJWVpaSknL69OlLly45\nOjrOmjWLtuw68UeC1VHV1dWEEI13O9LCpqamzp7Tcw8npRMoFIqff/55//79lZWVDg4Oy5cv\nZ6sQf12TSCSPHz/m8/kMwygUiqqqKoVCwTAMrUX8dUGhUMTGxlZUVLz//vv0IY/WIP66IBaL\nx44d6+npOXXqVFri7Ow8d+5cc3Pzb7/9NiEhYdKkSVZWVqQrxR8JVkfRZcmGhgb1qvr6etJK\nHg06hZOia5mZmfv27aNXBv38/MLDw5Wf4kH8dW3x4sX0DuuSkpL4+PjvvvuutLR04cKFtBbx\n14VTp07dunXLz8+P3ivdBsRfF8aOHTt27Fj18unTpyckJNTU1OTm5tIEq+vEH08RdpSZmRkh\nhO5mpoKuQ9JTDp0JJ0V36uvrd+/eHR0dnZub6+7uvnHjxsjISOXsiiD+ncjOzm7FihUMwyQn\nJ5eVldFCxJ9zeXl5hw4dsrOza+3JQWWIf2fi8/n0uUL2QdquE3+sYHWUo6NjYWGhRCJRKa+v\nr29sbCSEWFtb62NezzWcFB2prKxctWoV3cFv/vz5fn5+7GUpZYg/5yoqKo4ePcrj8cLDw1Wq\nzM3NraysJBJJQUEBvXUX8edcUVGRTCYrKSmZOXOmeu27775LCBkyZMjatWsJ4t/phEIhUbom\n2HXijxWsjvL09CSE3Lp1S6WclojFYpV/7qET4KToglQqXbduXWFh4eDBg3fv3u3v768xuyKI\nvw5079794sWL586dKy0tVamSyWR0pwYLCwtagvhzTiAQmKlh39FNTEzMzMyMjY3pt4g/56RS\n6eLFi8PDw4uLi9VrHz16RAhxdnam33ad+CPB6ij6+UfZ2dkqJ/7nn38mhIwYMaK1NyHQHZwU\nXUhLS8vLy+vTp8/q1atb28OaQvw5JxAIXF1dCSEpKSkqVZcuXWppaTEyMmLfYBB/zg0dOvSI\nmsOHD9PauLi4I0eORERE0G8Rf84JhUJ7e/vi4uLTp0+rVF26dEkikXTv3p39kMGuE38kWB3l\n4uIyfPhwhUIRExNTXl5OCGlubj58+HB6erpAIGB3xYTOhJOiCz/++CMhxN/fXyKRlGpCl98J\n4q8bNG4nT55MSkqi9+o2NDT8+OOPe/bsIYTMnDlTIPjvLR+Iv34h/roQFBTEMMyFCxcOHjxI\nb6Wqq6s7f/78zp07CSFz585lP46w68SfwZ79HVdVVRUZGUmTZUdHx/LycqlUyuPxli5dOm7c\nOH3P7plFd3J/6623NP7C4KRwS6FQzJgxg92rXaPly5ezj/kg/rqwffv21NRU+rWpqSlNswgh\no0aNioiI4PH+/x9mxL8TyOXy4OBgQsjevXvt7e2VqxB/XTh58uSBAwfo1+zPP8Mw06ZNe/vt\nt5XXpbpI/PkbNmzotMGeVcbGxhMmTKC70ZSVlZmYmAwcOHDp0qXqW/UDh65cuVJQUDBw4EBv\nb2/1WpwUblVUVCQkJLTdZsSIEfQyFkH8dWP48OEeHh41NTUKhaKurs7BwcHHx+edd94JCwtT\nueqB+HcChUJx9OhRQkhQUFC3bt2UqxB/XfD29vb19a2urpbJZHV1dY6Oji+++OLChQtffvnl\nrvnzjxUsAAAAAI7hHiwAAAAAjiHBAgAAAOAYEiwAAAAAjiHBAgAAAOAYEiwAAAAAjiHBAgAA\nAOAYEiwAAAAAjiHBAgAAAOAYEiwAAAAAjiHBAgAAAOAYEiwAAAAAjiHBAgAAAOAYEiwAAAAA\njiHBAgAAAOAYEiwAgOdRfHw8wzAMw+Tn5+uuS+dMDKALQoIFYHiamppqa2v/tllISAjDMCdO\nnOiEKQEAgDIkWACG5ODBg76+vt27dxeLxX379v3oo4+kUqnGlmlpaYmJif7+/jNmzOjkSQIA\nABIsAMOgUCjCwsJmz559586d5uZmhUKRk5Ozfv36wYMHV1dXqzSWy+UffPABj8fbvn27XmYL\nAPCcQ4IFYBg+++yzhIQEsVi8a9euoqIiiURy4sQJFxeXjIyMBQsWqDT+6quv7t69O3fu3EGD\nBulltgAAzzkkWACGYc+ePYSQb775ZuHChQ4ODlZWVmFhYcnJyQzDHDt2rKqqim1ZVVW1du1a\nsVj8ySef6G++AADPNSRYAAagsrLy4cOHhJCXX35ZudzLy8vZ2Vkul2dkZLCFGzduLCsrW7Nm\njZ2dXQfHzczMDA8Pd3d3NzExcXBwCAoK+umnn8j/nvOytLRU7yKXyw8cODBlypSePXsaGxv3\n7t170qRJx48fl8lk6o3d3NwYhqGJYGFh4cqVKz09PU1NTZ2cnMaPH3/48GGFQqFxYvX19R9/\n/PGwYcNsbGxEIpGTk9O0adMOHjyocZR+/fqxo1y5ciU0NNTBwUEkErm4uISHhxcXF9Nmf/zx\nx+zZs52cnOi0X3nlld9//13j6O16jYSQlpaW/fv3T548uW/fvqampm5ubhMmTNixY0dTU5PG\n9m148ODB+++/P2zYMDq0q6urn5/fxo0bS0pKWuty8+bNf/7zny7BJ0kwAAAOZElEQVQuLsbG\nxvb29oGBgX/73IMWXdobE+1GATAkCgAwBE5OToSQxMRE5cKcnBwej8cwjEQioSXZ2dlGRkbu\n7u5NTU0dHDE2NtbIyEjlLwbDMKtXr46LiyOEWFhYqHQpKCjw9fXV+Kemf//++fn5Ku1dXV0J\nITExMVevXrW1tSWEODo69unTh8f77/9+QUFBcrlcpdeNGzc8PDw0jjJw4MCCggKV9l5eXnSU\nHTt28Pl8kUjk4+NjZmZGuzg6OhYUFBw4cEAoFDIM4+npSWdCCOHz+SkpKR18jbm5ue7u7hrb\nu7i4PHr06OnPyObNm9nIqBCJRKmpqSrt5XL51q1b1U8iISQ0NDQ2NpZ+nZeX15EuWsREu1EA\nDAsSLADD8PHHHxNCzM3Nd+/eXVxcXFFRcerUKZqgzJgxg202ZcoUQsjp06c7ONzhw4fpmxyf\nzw8LC9u8efOmTZuCgoIYhiGE0LUrlQSroaGhd+/etFffvn1XrFixe/fuZcuWscmQo6NjUVGR\nchc6/4ULF9ra2o4ZMyY3N5eWFxQUTJw4kfbatWuXcpdbt24JBAJaNWLEiA8//DA+Pn7BggW9\nevWihV5eXvQhABZNsIYPH25kZLRp0yaaejY1NS1ZsoR26devH5/Pf+WVV0pLS2mX77//ns/n\n06qOvMba2lpvb29a5efnt2nTpn379sXExAwYMIAtfMozcu7cORp8Y2Pj2bNnx8bG0kP5+fnR\nQ1lYWFRVVSl3Wb16NZu4BAQErFmzJiYmJigoiAbQ3t5ePY/RoosW512LUQAMDhIsAMMgk8mm\nT5+u/h9///79KysraZvk5GRCyIQJEzo4VmNjo4uLCyHE2tpaZV3k5MmTJiYm7Du6chV7y9fS\npUuV188aGhree+89WjVr1izlLjTB4vF4fn5+KlmRRCKxsLAghAQFBSmXjxkzhnbZsmWLcnlF\nRcXUqVPpKDt37lSuogkWIWTDhg3K5S0tLezakr+/v8pS2dy5c2nVkydPtH6NZ86coYVRUVEq\nQ48bN45WqS/waBQSEkKzq/T0dJWqiIgIeqh///vfbGF+fr6xsTEhRCwWJyQkKLdPSUmxtrZm\nf4TYPEaLLlrERLtRAAwOEiwAQ/LNN9/079+fLq64ubmtX7++sbGRVjU3N9OVmIyMjA6Owl6j\n+fLLL9Vro6Oj1ROskpKS7t27E0LGjBmjflFPLpePHj2a9nr48CFbThMsQsjly5fVBwoICCCE\nuLu7syU//PCDxkSNqqmpoW/PNjY2dXV1bDlNsExNTRsaGlS6vP766/SAP//8s0rV119/Tatu\n3ryp9Wtcs2YNLcnKylJpf/bs2YCAgICAgGvXrqm/FnVOTk4CgSAkJES9KjMzk44SHx/PFs6Z\nM4cWqiwBUocOHVLPY7TookVMtBgFwBAhwQIwPPX19RUVFSqF9L6oRYsWdfz49PKcvb29xhu5\nSktL6d0zygnW7t276ZticnKyxmOyudHBgwfZQppg2dnZaewyc+ZMQoirqytbEhkZ2Vq+Qm3c\nuFElK1L8L8EaOnSoevvFixcTQhiGqa2tVak6ffo0PdT169e1fo1sqvrJJ59o7MKJ3377jY6i\nvHRHF+ecnZ2lUql6l5aWlj59+qjkMVp00SImWowCYIjwFCGA4TExMaGXz1gSiWTDhg2WlpYf\nffQRW5iXl/faa6/Z2NhYWVkFBwffu3fvKY+fk5NDCBkwYIBQKFSvtbW1ZVeeWNevXyeE8Hg8\nT09PiSbOzs605Z07d1T6sm+oTzmx7t27s1f9VAwZMoR+8eDBA5WqHj16tHZYMzMz9p73Nmjx\nGqdOnUqPvHr16lGjRsXGxt6+fbuNB+vapays7Pr16/Hx8W+88YZKlVQqzcvLI4QMGzZM473k\nfD5/xIgRHexC2h8T7UYBMEQCfU8AADiwfv36ioqKuLg49haWS5cuhYSElJeX02+TkpJSU1MP\nHTo0bdq0tg/V0NBAt4Rwc3NrrY2rq+v9+/eVSx4/fkwIkcvl7P3OrWGnxGJvav5bdNA2hmDn\nrDI9TmjxGvv06XP8+PHw8PBHjx5dvXr16tWrhBAzM7Nhw4ZNmTIlODj4bw+lrK6u7vvvv09N\nTc3Kynrw4EF9fX1rLXNzc2ka19rjloQQlWcbtehC2h8T7UYBMERYwQIweJmZmXv27PHy8lq0\naBEtKS0tnT59+pMnT957771ff/318uXLERERjY2Nr732Wm5ubttHozd1EULoM2saqa891NXV\nPeVsGxsbVUraGEiFXC5vuwE7MfVROk671zh58uTs7Ox9+/aFhITQ25Xq6upSU1NXrFjh6em5\nZMmS5ubmpznmhQsX3N3dw8PDjx07dvfuXYVC0b9//+nTp69bt059+yj2QYQ20Dv5OtKFtD8m\n2o0CYIiwggVg8JYtWyaTyWJjY9n9C6KioioqKj777LOoqChaMmrUqD59+oSHhy9btox9tE0j\nS0tLW1vbsrIyeilHo7/++ku9FyHEzMysurq6tb2aOs7T0zMrK0t9dBabPvbt25fz0bV+jaam\npvPmzZs3b55MJrt79+61a9dSUlLOnz/f2Nj4xRdfdOvWbdOmTW0fITs7Ozg4WCqVWlhYLF++\nPCQkxMvLi01E8vPzVdo7OzsLhUKpVKp+qZSlkmpr0YW0PybajQJgiLCCBWDYzpw5c/HixcmT\nJwcGBtIShUKRlJTE4/HYBS3q7bffNjU1TU5O/tvVHXqH0927d1taWtRra2pq1FOcfv36EULq\n6uraeOPsOHpdqbq6urVR/vOf/9AvdJFgafEai4qKCgsL2SuGfD7f19d30aJFCQkJWVlZ9LYw\ndsuxNuzdu1cqlRJCTp8+vXbtWh8fH+VlnidPnqi05/F49JrdjRs3NJ5EQsjNmzc72IW0Pyba\njQJgiJBgARgwqVS6YsUKIyOjbdu2sYWFhYUVFRXqjRmG4fF4zc3NEomk7cPSR+uLioqUn5ln\n7dy5Uz1FY/egop+ZqG7btm0ODg4ODg5tLIz9LXZTzZiYGPXa2tpa+tSeubk5u70nh7R4jePH\nj3dychowYIBC7TN/XF1d6S35lZWVfzs0PaBAIPD391evPXfunHrh+PHjCSH5+fnsfhPKEhMT\n//jjj4530S4m7R0FwCDp9yFGAOiIzZs3E0KWLVumXCiXy+mt7p9//rly+f79+wkh1tbWf3vY\nyspKGxsbQkiPHj2uXLmiXPXTTz+JxWL610Nlo9Hg4GBCCMMwBw4cUDngL7/8Qm8/CggIUC6n\nTyOGhYVpnIb6Ng0KhWLUqFGEED6f/8UXX6jMOSgoiE4sNjZWuYouyE2dOlV9CLpNQ7du3dSr\n1Ldp0OI1zpo1ix5k3759Ku2zsrJoJAMDAzW+fGXs/hS//fabSlVSUhL7sOemTZvY8qKiIlNT\nU0KIubm5ys7+169fd3R0ZN8F2N0QtOiiRUy0GwXA4CDBAjBUJSUlYrHYxsZGfU8susDD5/OX\nLl169erV9PT06Oho+jassst5a9gFCSMjozfeeCM2Nnbbtm2vvvoqj8czMjKiqUyPHj2Uu+Tl\n5dE3TkLIyJEj161bt3///u3bt4eGhtKbw2xsbO7du6fcRYsEKz09nb3XZ8yYMevWrdu7d+/7\n779Pt54nhPTr109lgyUOE6z2vsa0tDQ6Wz6fHxoaumPHjgMHDuzYsWPOnDl0+wYej3f27NlW\nT4PSceijAA4ODlu2bLl48eKPP/64a9cuuh28tbU1vWLYs2fPLVu2PHjwgPZav349nSrDMIGB\ngRs2bNiyZcuMGTPoTwLdE18lj9GiixbnXYtRAAwOEiwAQzV//nxCyO7du9Wr5HI5u1+2snnz\n5j398fft20c/0kSZWCxOSkqiK2dubm4qXTIyMgYNGqQ+LiGkd+/eypt/UlokWAqF4tSpU3Rd\nRF1YWFh5eblKew4TLC1e45YtW1p7LM7ExETjXvkasRvoq3jxxRfv37/PfrAPIeTo0aO0i0wm\nYz+sRsX06dPZ/d+V8xgtumgRE+1GATAsSLAADNLt27d5PN4LL7zQ0tLSWpvz589PnTq1Z8+e\ndnZ2kydP/uGHH9o7yp07d9544w1HR0ehUNirV693332Xfh4zzUsGDRqk3qW5uXn79u2jR4+2\ntbUViUQeHh6TJ0/et2+fykcNUtolWAqForS0dN26dYMHD7a0tDQyMnJwcJg6derx48c1Hofb\nBKu9r1GhUNy/f3/BggVDhw51dHQUiURubm5jx45dt24d+9nST+nXX3+dPn26h4eHSCSyt7ef\nMmXKoUOHZDKZQqFobGyMiYmZOHHijBkzfv/9d+Vely9ffvPNN52dnUUikbW19ejRo7/77juF\nQlFeXr5y5cqVK1cqf9ii1l3aGxPtRgEwIIxC7dZLAIC2BQUFnT179vXXX3+aJ+C0Vl9f//vv\nv//111/qO5V3ArlcnpOTc/v27TFjxijfGAQA8DSQYAGAqpMnTyYlJRFC4uLirKysVGofP37c\nu3fvhoaGzZs3R0RE6GOCAABdHTYaBQBVYrGYbtDg4+Pz4YcfKlcpFIo1a9Y0NDQQQkJCQvQz\nPwCALg8rWACgqqmpydfXNysri2GYlStXzpkzx83Nraqq6t69ezExMWlpaYSQWbNmadwlCwAA\nCBIsANAoLy9v5MiRxcXFGmuHDBmSnJysfvUQAAAo7OQOABq4urrm5ORs3rzZy8tLJBLRQvpJ\nL1u3bv3ll1+QXQEAtAErWADwNxQKxePHjwkhVlZWbLIFAABtQIIFAAAAwDFcIgQAAADgGBIs\nAAAAAI4hwQIAAADgGBIsAAAAAI4hwQIAAADgGBIsAAAAAI4hwQIAAADgGBIsAAAAAI4hwQIA\nAADgGBIsAAAAAI4hwQIAAADgGBIsAAAAAI4hwQIAAADgGBIsAAAAAI4hwQIAAADgGBIsAAAA\nAI79H/pumCuscBKpAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 200,
       "width": 400
      },
      "text/plain": {
       "height": 200,
       "width": 400
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plotting\n",
    "p = bmk_update_s %>%\n",
    "    ggplot(aes(perc_added, cpu_hour_perc_savings)) +\n",
    "    geom_smooth(method='lm') +\n",
    "    geom_point() +\n",
    "    labs(x='% genomes added', y='% CPU hours saved') +\n",
    "    theme_bw()\n",
    "\n",
    "p.dims(4,2)\n",
    "plot(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# sessionInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T07:40:16.290393Z",
     "start_time": "2021-02-07T07:40:15.404Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Struo2\n",
      "======\n",
      "\n",
      "**Struo2:** a pipeline for building custom databases for common metagenome profilers\n",
      "\n",
      "> \"Struo\" --> from the Latin: “I build” or “I gather”\n",
      "\n",
      "![](https://media.giphy.com/media/lPSMFWqKCJY5Let2v5/giphy.gif)\n",
      "\n",
      "* Version: 2.1.2\n",
      "\n",
      "--- conda envs ---\n",
      "==> /ebio/abt3_projects2/databases_no-backup/bin/struo2_1/bin/envs/genes.yaml <==\n",
      "channels:\n",
      "- bioconda\n",
      "dependencies:\n",
      "- pigz\n",
      "- python=3\n",
      "- numpy\n",
      "- bioconda::seqkit\n",
      "- bioconda::fasta-splitter\n",
      "- bioconda::vsearch\n",
      "- bioconda::prodigal\n",
      "- bioconda::mmseqs2\n",
      "==> /ebio/abt3_projects2/databases_no-backup/bin/struo2_1/bin/envs/humann2.yaml <==\n",
      "channels:\n",
      "- bioconda\n",
      "dependencies:\n",
      "- pigz\n",
      "- bioconda::bowtie2\n",
      "- bioconda::vsearch\n",
      "- bioconda::prodigal\n",
      "- bioconda::diamond=0.8.36\n",
      "\n",
      "==> /ebio/abt3_projects2/databases_no-backup/bin/struo2_1/bin/envs/humann3.yaml <==\n",
      "channels:\n",
      "- bioconda\n",
      "- biobakery\n",
      "dependencies:\n",
      "- pigz\n",
      "- bioconda::seqkit\n",
      "- bioconda::vsearch\n",
      "- bioconda::prodigal\n",
      "- bioconda::diamond=0.9.24\n",
      "- biobakery::humann\n",
      "\n",
      "==> /ebio/abt3_projects2/databases_no-backup/bin/struo2_1/bin/envs/kraken2.yaml <==\n",
      "channels:\n",
      "- bioconda\n",
      "dependencies:\n",
      "- libiconv\n",
      "- bioconda::kraken2\n",
      "- bioconda::bracken\n"
     ]
    }
   ],
   "source": [
    "pipelineInfo('/ebio/abt3_projects2/databases_no-backup/bin/struo2_1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-25T13:12:35.952378Z",
     "start_time": "2021-01-25T13:12:11.187Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 3.6.3 (2020-02-29)\n",
       "Platform: x86_64-conda_cos6-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 18.04.5 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /ebio/abt3_projects/Georg_animal_feces/envs/tidyverse/lib/libopenblasp-r0.3.9.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       "[1] LeyLabRMisc_0.1.6 tidytable_0.5.1   data.table_1.12.8 ggplot2_3.3.1    \n",
       "[5] tidyr_1.1.0       dplyr_1.0.0      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] Rcpp_1.0.4.6     pillar_1.4.4     compiler_3.6.3   base64enc_0.1-3 \n",
       " [5] tools_3.6.3      digest_0.6.25    uuid_0.1-4       jsonlite_1.6.1  \n",
       " [9] evaluate_0.14    lifecycle_0.2.0  tibble_3.0.1     gtable_0.3.0    \n",
       "[13] pkgconfig_2.0.3  rlang_0.4.6      IRdisplay_0.7.0  cli_2.0.2       \n",
       "[17] IRkernel_1.1     repr_1.1.0       withr_2.2.0      generics_0.0.2  \n",
       "[21] vctrs_0.3.1      grid_3.6.3       tidyselect_1.1.0 glue_1.4.1      \n",
       "[25] R6_2.4.1         fansi_0.4.1      pbdZMQ_0.3-3     farver_2.0.3    \n",
       "[29] purrr_0.3.4      magrittr_1.5     scales_1.1.1     ellipsis_0.3.1  \n",
       "[33] htmltools_0.4.0  assertthat_0.2.1 colorspace_1.4-1 labeling_0.3    \n",
       "[37] munsell_0.5.0    crayon_1.3.4    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:Georg_animal_feces-tidyverse]",
   "language": "R",
   "name": "conda-env-Georg_animal_feces-tidyverse-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  },
  "notify_time": "30",
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
