# method for creating db
# Use UniRef-IDs for gene clusters
# For each gene cluster
  # Taxonomic breadth: specific to any particular species (allowing for N non-species)?
    # species determined by genome taxonomy
    # if species-specific:
      # add to metaphlan database
      # add marker sequence (w/ correct ID) to marker fasta database
      
rule metaphlan3_species_specific_markers:
    """
    Adding new species-specific markers to the metaphlan pkl database 
    """
    input:
        pkl = config['params']['metaphlan3']['pkl'],
        mp_fna = config['params']['metaphlan3']['fasta'],
        names = config['names_dmp'],
        fna = humann3_dir + '{uniref}/genome_reps_filt_annot.fna.gz',
        tsv = humann3_dir + '{uniref}/genome_reps_filt_annot.tsv.gz',
    output:
        tsv = config['tmp_dir'] + '{uniref}/metaphlan3/species_specific.txt'
    params:
        params = config['params']['metaphlan3']['species_specific_markers'],
        exe = config['pipeline']['script_folder'] + 'metaphlan_db_from_uniref.py'
    resources:
        time = lambda wildcards, attempt: attempt ** 3 * 59,
        mem_gb_pt = lambda wildcards, attempt: attempt * 2 * 8
    conda:
        '../../envs/genes.yaml'
    log:
        log_dir + 'metaphlan3_species_specific_markers/{uniref}.log'
    benchmark:
        benchmark_dir + 'metaphlan3_species_specific_markers/{uniref}.txt'
    shell:
        """
        {params.exe} {params.params} \
           {input.fna} {input.tsv} \
           {input.pkl} {input.mp_fna} \
           {input.names} 2> {log} 
        """

